// @sap-ux/cds-annotation-parser/dist/transformer/annotation-ast-nodes.js
import { Range } from '@sap-ux/text-document-utils';
import { copyRange } from './range.js';
export var Delimiter;
(function (Delimiter) {
    Delimiter["none"] = "none";
    Delimiter["quoted"] = "quoted";
    Delimiter["exclamationSquareBrackets"] = "exclamation-square-brackets";
})(Delimiter || (Delimiter = {}));
export const TOKEN_TYPE = 'token';
export const NUMBER_LITERAL_TYPE = 'number';
export const ENUM_TYPE = 'enum';
export const QUOTED_LITERAL_TYPE = 'quoted-literal';
export const EMPTY_VALUE_TYPE = 'empty-value';
export const BOOLEAN_TYPE = 'boolean';
export const STRING_LITERAL_TYPE = 'string';
export const MULTI_LINE_STRING_LITERAL_TYPE = 'multi-line-string';
export const PATH_TYPE = 'path';
export const SEPARATOR_TYPE = 'separator';
export const IDENTIFIER_TYPE = 'identifier';
export const OPERATOR_TYPE = 'operator';
export const UNSUPPORTED_OPERATOR_EXPRESSION_TYPE = 'unsupported-operator-expression';
export const INCORRECT_EXPRESSION_TYPE = 'incorrect-expression';
export const CORRECT_EXPRESSION_TYPE = 'correct-expression';
export const EXPRESSION_TYPES = [
    UNSUPPORTED_OPERATOR_EXPRESSION_TYPE,
    INCORRECT_EXPRESSION_TYPE,
    CORRECT_EXPRESSION_TYPE
];
export const RECORD_PROPERTY_TYPE = 'record-property';
export const RECORD_TYPE = 'record';
export const COLLECTION_TYPE = 'collection';
export const QUALIFIER_TYPE = 'qualifier';
export const ANNOTATION_TYPE = 'annotation';
export const ANNOTATION_GROUP_TYPE = 'annotation-group';
export const ANNOTATION_GROUP_ITEMS_TYPE = 'annotation-group-items';
const CONTAINER_TYPES = new Set([ANNOTATION_GROUP_ITEMS_TYPE, RECORD_TYPE, COLLECTION_TYPE]);
export const isContainer = (node) => CONTAINER_TYPES.has(node.type);
export const nodeRange = (node, includeDelimiters) => {
    if (!node.range) {
        return undefined;
    }
    if (includeDelimiters) {
        return copyRange(node.range);
    }
    switch (node.type) {
        case ANNOTATION_GROUP_ITEMS_TYPE:
        case RECORD_TYPE:
        case COLLECTION_TYPE:
        case STRING_LITERAL_TYPE:
        case INCORRECT_EXPRESSION_TYPE:
        case UNSUPPORTED_OPERATOR_EXPRESSION_TYPE:
        case CORRECT_EXPRESSION_TYPE: {
            // if delimiter tokens exist adjust range accordingly
            const start = node.openToken?.range ? copyRange(node.openToken.range).end : copyRange(node.range).start;
            const end = node.closeToken?.range ? copyRange(node.closeToken.range).start : copyRange(node.range).end;
            return Range.create(start, end);
        }
        default:
            break;
    }
    return copyRange(node.range);
};
