// @sap-ux/eslint-plugin-fiori-tools/lib/project-context/project-context.js
import { existsSync, readFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import { createSyncFn } from 'synckit';
import { normalizePath } from '@sap-ux/project-access';
import { ApplicationParser } from './parser/index.js';
import { DiagnosticCache } from '../language/diagnostic-cache.js';
import { linkProject } from './linker/index.js';
// Sync function for worker calls
let artifactWorker;
/**
 * Gets the file system path to the worker script.
 *
 * @param name - The name of the worker file
 * @returns The absolute path to the worker file
 */
function getWorkerPath(name) {
    // Create sync function to call the working draft-toggle-worker
    const currentDir = dirname(fileURLToPath(import.meta.url));
    // Try multiple possible worker locations
    const workerPaths = [
        join(currentDir, name), // src location
        join(currentDir, '..', '..', 'lib', 'project-context', name) // dist location
    ];
    let workerPath = null;
    for (const tryPath of workerPaths) {
        if (existsSync(tryPath)) {
            workerPath = tryPath;
            break;
        }
    }
    if (!workerPath) {
        throw new Error(`Worker not found in any expected location: ${workerPaths.join(', ')}`);
    }
    return workerPath;
}
/**
 * Gets or creates the artifact worker instance for finding Fiori artifacts.
 *
 * @returns A synchronous function that takes a file path and returns worker results
 */
function getArtifactWorker() {
    if (artifactWorker) {
        return artifactWorker;
    }
    const workerPath = getWorkerPath('artifacts.js');
    artifactWorker = createSyncFn(workerPath, { timeout: 10000 });
    return artifactWorker;
}
/**
 * Manages the project context including parsed artifacts, indexes, and linked models.
 * Provides access to project structure, manifests, and services.
 */
export class ProjectContext {
    artifacts;
    _index;
    /**
     * Gets the parsed project index containing all project information.
     */
    get index() {
        return this._index;
    }
    _linkedModel;
    /**
     * Gets the linked model containing resolved references between project entities.
     */
    get linkedModel() {
        return this._linkedModel;
    }
    documents = {};
    /**
     * Creates a new ProjectContext instance.
     *
     * @param artifacts - The found Fiori artifacts in the project
     * @param index - The parsed project index
     * @param linkedModel - The linked model with resolved references
     */
    constructor(artifacts, index, linkedModel) {
        this._index = index;
        this._linkedModel = linkedModel;
        this.artifacts = artifacts;
    }
    /**
     * Gets the indexed service for the main service of an application.
     *
     * @param appIndex - The parsed application index
     * @param serviceName - Optional service name, defaults to the main service name from manifest
     * @returns The parsed service or undefined if not found
     */
    getIndexedServiceForMainService(appIndex, serviceName) {
        const name = serviceName ?? appIndex.manifest.mainServiceName;
        return appIndex.services[name];
    }
    /**
     * Gets the manifest for a specific application.
     *
     * @param app - Optional application key, defaults to the first application in the index
     * @returns The manifest object or undefined if not found
     */
    getManifest(app) {
        const key = app ?? Object.keys(this.index.apps)[0];
        const appIndex = this.index.apps[key];
        if (!appIndex) {
            return;
        }
        return appIndex.manifestObject;
    }
    /**
     * Re-indexes the project after a file has been updated.
     *
     * @param uri - The URI of the file to reindex
     * @param content - The updated content of the file
     */
    reindex(uri, content) {
        ProjectContext.fileCache.set(uri, content);
        const { diagnostics, index } = ProjectContext.parser.reparse(uri, this.index, ProjectContext.fileCacheProxy);
        for (const diagnostic of diagnostics) {
            DiagnosticCache.addMessage(uri, diagnostic.type, diagnostic);
        }
        const [linkedModel, linkerDiagnostics] = linkProject(index);
        for (const diagnostic of linkerDiagnostics) {
            DiagnosticCache.addMessage(uri, diagnostic.type, diagnostic);
        }
        this._linkedModel = linkedModel;
    }
    static parser = new ApplicationParser();
    /**
     * Project file mapping to artifacts. Used to find out to which project the file belongs.
     * It should only be used by `findProjectRoot` method.
     */
    static projectArtifactCache = new Map();
    static instanceCache = new Map();
    static updateCache = new Map();
    static appRoots = new Set();
    /**
     * Finds Fiori artifacts in the project using a worker process.
     *
     * @param _uri - The URI to start searching from (currently unused, uses process.cwd())
     * @returns The worker result containing found artifacts and project type
     */
    static findFioriArtifacts(_uri) {
        // potential issue when called from application modeler or via ESLint API
        const root = normalizePath(process.cwd());
        try {
            const cachedValue = this.projectArtifactCache.get(root);
            if (cachedValue) {
                return cachedValue;
            }
            const artifacts = getArtifactWorker()(root);
            this.projectArtifactCache.set(root, artifacts);
            return artifacts;
        }
        catch (error) {
            console.error('Error finding Fiori artifacts:', error);
            return { artifacts: {}, projectType: 'EDMXBackend' };
        }
    }
    /**
     * If set to true, forces re-indexing on the first update of a file.
     */
    static forceReindexOnFirstUpdate = false; // NOSONAR - Property must be mutable for test setup
    /**
     * Creates a ProjectContext for the given file path.
     *
     * @param uri - The URI of the file to get the context for
     * @returns A ProjectContext instance
     */
    static getInstanceForFile(uri) {
        const cachedValue = this.instanceCache.get(uri);
        if (cachedValue) {
            return cachedValue;
        }
        return this.createForFile(uri);
    }
    /**
     * Updates a file in the cache and triggers re-indexing if necessary.
     *
     * @param uri - The URI of the file to update
     * @param content - The new content of the file
     * @returns The ProjectContext instance for the file
     */
    static updateFile(uri, content) {
        this.fileCache.set(uri, content);
        const numberOfUpdates = this.updateCache.get(uri) ?? 0;
        this.updateCache.set(uri, numberOfUpdates + 1);
        const context = this.getInstanceForFile(uri);
        if (numberOfUpdates === 0 && !this.forceReindexOnFirstUpdate) {
            // assume that first time check is called the file content is the same as in file system/cache
            return context;
        }
        context.reindex(uri, content);
        return context;
    }
    static fileCache = new Map();
    static fileCacheProxy = new Proxy(this.fileCache, {
        get: (target, prop) => {
            if (prop === 'get') {
                return (key) => {
                    if (this.fileCache.has(key)) {
                        return this.fileCache.get(key);
                    }
                    try {
                        const path = fileURLToPath(key);
                        const content = readFileSync(path, 'utf-8');
                        this.fileCache.set(key, content);
                        return content;
                    }
                    catch {
                        return undefined;
                    }
                };
            }
        }
    });
    /**
     * Creates a ProjectContext for the given file path by parsing artifacts and building indexes.
     *
     * @param uri - The URI of the file to create the context for
     * @returns A new ProjectContext instance
     */
    static createForFile(uri) {
        // not all project files might be indexed, but eslint still will pick it up
        // try to assign the file to the closest known app root and avoid re-indexing
        for (const appRoot of this.appRoots.values()) {
            if (uri.startsWith(appRoot)) {
                const cachedValue = this.instanceCache.get(appRoot);
                if (cachedValue) {
                    this.instanceCache.set(uri, cachedValue);
                    return cachedValue;
                }
            }
        }
        const { artifacts, projectType } = this.findFioriArtifacts(uri);
        const { diagnostics, index } = this.parser.parse(projectType, artifacts, this.fileCacheProxy);
        for (const diagnostic of diagnostics) {
            DiagnosticCache.addMessage(uri, diagnostic.type, diagnostic);
        }
        const [linkedModel, linkerDiagnostics] = linkProject(index);
        for (const diagnostic of linkerDiagnostics) {
            DiagnosticCache.addMessage(uri, diagnostic.type, diagnostic);
        }
        const context = new ProjectContext(artifacts, index, linkedModel);
        for (const uri of Object.keys(index.documents)) {
            this.instanceCache.set(uri, context);
        }
        for (const appRoot of Object.keys(index.apps)) {
            this.instanceCache.set(appRoot, context);
            this.appRoots.add(appRoot);
        }
        return context;
    }
}
