// @sap-ux/logger/dist/winston-logger/logger.js
import { ConsoleTransport } from '../transports/index.js';
import { LogLevel } from '../types.js';
import winston, { format } from 'winston';
import { toWinstonLogLevel, toWinstonTransport } from './adapter.js';
import { nextColor } from './utils.js';
import { inspect } from 'node:util';
const defaultLoggerOptions = {
    transports: [new ConsoleTransport()]
};
class BaseWinstonLogger {
    _logger;
    logPrefix;
    logPrefixColor;
    winstonLevel;
    metadataOverride;
    // Maintain of map of transports. This is useful for adding/removing transports
    transportMap;
    initialize({ logger, transportMap, metadataOverride, winstonLevel, logPrefix }) {
        this._logger = logger;
        this.transportMap = transportMap;
        this.winstonLevel = winstonLevel;
        this.logPrefix = logPrefix;
        this.metadataOverride = metadataOverride;
    }
    info(message) {
        this.log({ level: LogLevel.Info, message });
    }
    warn(message) {
        this.log({ level: LogLevel.Warn, message });
    }
    error(message) {
        this.log({ level: LogLevel.Error, message });
    }
    debug(message) {
        this.log({ level: LogLevel.Debug, message });
    }
    log(data) {
        if (!this.transportMap.size) {
            // Nothing to do
            return;
        }
        if (typeof data === 'string') {
            this.winstonLog({ level: this.winstonLevel, message: data, metadata: this.metadataOverride });
        }
        else {
            const level = toWinstonLogLevel(data.level) ?? this._logger.level;
            this.winstonLog({ level, message: data.message, metadata: this.metadataOverride });
        }
    }
    winstonLog({ level, message, metadata }) {
        const msg = typeof message === 'string' ? message : inspect(message);
        this._logger.log(level, msg, metadata);
    }
    addToMap(transportMap, transport) {
        const winstonTransport = toWinstonTransport(transport);
        if (!transportMap.has(transport)) {
            transportMap.set(transport, winstonTransport);
            return winstonTransport;
        }
        return undefined;
    }
    add(transport) {
        const winstonTransport = this.addToMap(this.transportMap, transport);
        if (winstonTransport) {
            this._logger.add(winstonTransport);
        }
        return this;
    }
    remove(transport) {
        const winstonTransport = this.transportMap.get(transport);
        if (winstonTransport) {
            this._logger.remove(winstonTransport);
            this.transportMap.delete(transport);
            return this;
        }
        else {
            throw new Error('Cannot remove non-existent transport');
        }
    }
    transports() {
        return Array.from(this.transportMap.keys());
    }
    child({ logPrefix }) {
        const childLogPrefix = `${this.logPrefix}.${logPrefix}`;
        const metadataOverride = { label: childLogPrefix, labelColor: nextColor() };
        const childWinstonLogger = this._logger.child(metadataOverride);
        const childLogger = new BaseWinstonLogger();
        childLogger.initialize({
            logger: childWinstonLogger,
            transportMap: this.transportMap,
            winstonLevel: this.winstonLevel,
            logPrefix: childLogPrefix,
            metadataOverride
        });
        return childLogger;
    }
}
/**
 *  Winston implementation of the @type {Logger} interface
 */
export class WinstonLogger extends BaseWinstonLogger {
    constructor({ logLevel = LogLevel.Info, transports = [], logPrefix = 'main' } = defaultLoggerOptions) {
        super();
        const transportMap = new Map();
        transports.forEach((t) => this.addToMap(transportMap, t));
        const level = toWinstonLogLevel(logLevel);
        const logger = winston.createLogger({
            level,
            transports: Array.from(transportMap.values()),
            format: format.combine(format.timestamp(), format.json(), format.splat()),
            defaultMeta: { label: logPrefix, labelColor: nextColor() }
        });
        const winstonLevel = level ?? logger.level;
        this.initialize({ logger, transportMap, winstonLevel, logPrefix });
    }
}
