// @sap-ux/fiori-annotation-api/dist/cds/adapter.js
import { fileURLToPath, pathToFileURL } from 'node:url';
import { basename, dirname, join, relative, sep } from 'node:path';
import { TextDocument } from 'vscode-languageserver-textdocument';
import cdsCompilerFacade from '@sap/ux-cds-compiler-facade';
const { createCdsCompilerFacadeForRoot, createMetadataCollector, getMetadataElementsFromMap } = cdsCompilerFacade;
import { TEXT_TYPE, ELEMENT_TYPE, DiagnosticSeverity, createAttributeNode, createElementNode, createTextNode, ANNOTATION_FILE_TYPE, Range } from '@sap-ux/odata-annotation-core-types';
import { MetadataService } from '@sap-ux/odata-entity-model';
import { ANNOTATION_GROUP_ITEMS_TYPE, ANNOTATION_GROUP_TYPE, ANNOTATION_TYPE, COLLECTION_TYPE, QUALIFIER_TYPE, RECORD_PROPERTY_TYPE, RECORD_TYPE, isReservedProperty } from '@sap-ux/cds-annotation-parser';
import { Edm, getAliasInformation, getAllNamespacesAndReferences, getElementAttribute, getElementAttributeValue, isElementWithName, parseIdentifier, toFullyQualifiedName } from '@sap-ux/odata-annotation-core';
import { TARGET_TYPE, printTarget } from '@sap-ux/cds-odata-annotation-converter';
import { convertTargets } from '../sap/index.js';
import { logger } from '../logger.js';
import { INSERT_TARGET, DELETE_ELEMENT, INSERT_ELEMENT, INSERT_ATTRIBUTE, DELETE_ATTRIBUTE, UPDATE_ATTRIBUTE_VALUE } from '../types/index.js';
import { ApiError, ApiErrorCode } from '../error.js';
import { CDSWriter } from './writer.js';
import { getMissingRefs } from './references.js';
import { addAllVocabulariesToAliasInformation } from '../vocabularies.js';
import { CDS_DOCUMENT_TYPE, getDocument, getGhostFileDocument } from './document.js';
import { convertPointer, getAstNodesFromPointer } from './pointer.js';
import { getGenericNodeFromPointer, pathFromUri, PRIMITIVE_TYPE_NAMES } from '../utils/index.js';
import { INSERT_PRIMITIVE_VALUE_TYPE, INSERT_TARGET_CHANGE_TYPE, createDeleteAnnotationChange, createDeleteQualifierChange, createInsertCollectionChange, createInsertEmbeddedAnnotationChange, createInsertPrimitiveValueChange, createInsertRecordChange, createInsertRecordPropertyChange, createInsertReferenceChange, createInsertTargetChange, createUpdatePrimitiveValueChange } from './change.js';
import { DELETE_REFERENCE, MOVE_ELEMENT, REPLACE_ATTRIBUTE, REPLACE_ELEMENT, REPLACE_ELEMENT_CONTENT, REPLACE_TEXT, UPDATE_ELEMENT_NAME } from '../types/internal-change.js';
/**
 *
 */
export class CDSAnnotationServiceAdapter {
    service;
    project;
    vocabularyService;
    appName;
    writeSapAnnotations;
    ignoreChangedFileInitialContent;
    metadataService = new MetadataService();
    splitAnnotationSupport = true;
    fileCache;
    documents = new Map();
    metadata = [];
    missingReferences = {};
    /**
     *
     * @param service - CDS service structure.
     * @param project - Project structure.
     * @param vocabularyService - Vocabulary API.
     * @param appName - Name of the application.
     * @param writeSapAnnotations - If set to true will write SAP annotations instead of OData annotations.
     * @param ignoreChangedFileInitialContent Flag indicating if to be changed files can be treated as empty.
     */
    constructor(service, project, vocabularyService, appName, writeSapAnnotations, ignoreChangedFileInitialContent) {
        this.service = service;
        this.project = project;
        this.vocabularyService = vocabularyService;
        this.appName = appName;
        this.writeSapAnnotations = writeSapAnnotations;
        this.ignoreChangedFileInitialContent = ignoreChangedFileInitialContent;
        this.fileCache = new Map();
        this._fileSequence = service.serviceFiles;
    }
    facade;
    setFileCache(fileCache) {
        this.fileCache = fileCache;
    }
    setFacade(facade) {
        this.facade = facade;
    }
    _compiledService;
    /**
     * @returns Compiled CDS service.
     */
    get compiledService() {
        if (!this._compiledService) {
            this._compiledService = this._getCompiledService();
        }
        return this._compiledService;
    }
    set compiledService(v) {
        this._compiledService = v;
    }
    _fileSequence;
    /**
     * Refreshes internal data structures from the provided project files.
     *
     * @param fileCache - File uri mapped to file content.
     * @returns Sync errors
     */
    async sync(fileCache) {
        const paths = [...this.service.serviceFiles.map((file) => fileURLToPath(file.uri))];
        const facade = await createCdsCompilerFacadeForRoot(this.project.root, paths, fileCache);
        this.setFacade(facade);
        const compileErrors = facade.getCompilerErrors(this.project.root);
        const relevantErrors = [...compileErrors].filter(([file, compilerMessage]) => {
            return (compilerMessage.messages.filter((value) => value.severity === DiagnosticSeverity.Error).length &&
                !file.startsWith('../'));
        });
        if (relevantErrors.length > 0) {
            // if model has compiler errors
            for (const [relativePath, compilerMessage] of relevantErrors) {
                logger.log(`Compile errors in: ${relativePath}`);
                for (const [fileUri, content] of fileCache) {
                    if (fileUri.endsWith(relativePath)) {
                        logger.log(content);
                    }
                }
                logger.log(JSON.stringify(compilerMessage, undefined, 2));
            }
            return compileErrors;
        }
        this.documents.clear();
        this.invalidateCaches();
        this.updateFileSequence(facade);
        this.setFileCache(fileCache);
        const metadataElementMap = facade.getMetadata(this.service.serviceName);
        // We collect already full metadata from compile model, we don't need to build it based on paths.
        const metadataCollector = createMetadataCollector(new Map(), facade);
        const { propagationMap, sourceUris } = facade.getPropagatedTargetMap(this.service.serviceName);
        for (const file of this.service.serviceFiles) {
            const document = getDocument(this.service.serviceName, this.vocabularyService, facade, fileCache, file, metadataCollector);
            this.documents.set(file.uri, document);
            // ghost files
            if (sourceUris.has(relative(this.project.root, fileURLToPath(file.uri)))) {
                const ghostDoc = getGhostFileDocument(this.service.serviceName, this.vocabularyService, facade, fileCache, file, metadataCollector, propagationMap);
                this.documents.set(ghostDoc.annotationFile.uri, ghostDoc);
            }
        }
        const metadataElements = getMetadataElementsFromMap(metadataElementMap);
        this.metadataService = new MetadataService({ uriMap: facade?.getUriMap() || new Map() });
        this.metadataService.import(metadataElements, 'DummyMetadataFileUri');
        return new Map();
    }
    /**
     * Get annotation documents.
     *
     * @returns Annotation documents.
     */
    getDocuments() {
        const annotationFiles = {};
        for (const [uri, document] of this.documents.entries()) {
            annotationFiles[uri] = document.annotationFile;
        }
        return annotationFiles;
    }
    /**
     * Returns all relevant service files.
     *
     * @param includeGhostFiles - Flag indicating if ghost files should be included.
     * @returns All relevant service files.
     */
    getAllFiles(includeGhostFiles) {
        if (includeGhostFiles) {
            const annotationFiles = [...this.compiledService.annotationFiles];
            annotationFiles.reverse();
            return annotationFiles.map((file) => ({
                uri: file.uri,
                isReadOnly: 
                // ghost files are readOnly and should not be in service files
                this.service.serviceFiles.find((serviceFile) => serviceFile.uri === file.uri)?.isReadOnly ?? true
            }));
        }
        return [...this.service.serviceFiles];
    }
    /**
     * Creates empty annotation file content for the given service.
     *
     * @param serviceName - Name of the service.
     * @param uri - URI for the new annotation file.
     * @returns New annotation file content.
     */
    getInitialFileContent(serviceName, uri) {
        if (this.facade) {
            const fileName = this.facade.getFileName(serviceName) ?? '';
            let path = relative(dirname(uri), join(this.project.root, dirname(fileName))).replace(/\\/g, '/');
            path = join(path, basename(fileName, '.cds'));
            path = path.split(sep).join('/'); // always use '/' instead of platform specific separator
            return `using ${serviceName} as service from '${path}';\n`;
        }
        return '';
    }
    /**
     * Converts changes to workspace edits.
     *
     * @param changes - Internal changes.
     * @returns Workspace edits.
     */
    async getWorkspaceEdit(changes) {
        const workspaceChanges = {};
        this.clearState();
        const writers = new Map();
        if (this.writeSapAnnotations) {
            this.handleSapAnnotations(writers, changes);
        }
        else {
            for (const change of changes) {
                const document = this.documents.get(change.uri);
                if (!document || !this.facade) {
                    continue;
                }
                const writer = this.getWriterForChange(writers, change);
                const changeHandler = this[change.type];
                changeHandler(writer, document, change);
            }
        }
        for (const [uri, writer] of writers.entries()) {
            const document = this.documents.get(uri);
            if (!document && !this.ignoreChangedFileInitialContent) {
                continue;
            }
            this.processMissingReferences(uri, writer);
            const edits = await writer.getTextEdits();
            workspaceChanges[uri] = edits;
        }
        return {
            changes: workspaceChanges
        };
    }
    /**
     * Returns a map of value list references.
     *
     * @returns Map of value list references.
     */
    getValueListReferences() {
        return new Map();
    }
    /**
     * Refreshes internal data structures from the provided external service file.
     *
     * @param _uri - URI of the external service metadata file.
     * @param _data - Metadata file content.
     */
    syncExternalService(_uri, _data) {
        // not supported in cds
    }
    /**
     *
     * @returns
     */
    getExternalServices() {
        return [];
    }
    handleSapAnnotations(writers, changes) {
        if (changes.length === 0) {
            return;
        }
        // only writing to a single files is supported
        const uri = changes[0].uri;
        const writer = this.getWriterForChange(writers, changes[0]);
        const targets = convertTargets({
            [uri]: changes
                .map((change) => {
                if (change.type === INSERT_TARGET) {
                    const annotationFile = this.documents.get(change.uri)?.annotationFile ?? this.createEmptyAnnotationFile(change);
                    const aliasInfo = getAliasInfo(annotationFile, this.metadataService, this.vocabularyService);
                    const missingReferences = getMissingRefs(this.documents, change.uri, change.target.name, change.target, aliasInfo, this.metadataService, {
                        apps: Object.keys(this.project.apps),
                        projectRoot: this.project.root,
                        appName: this.appName
                    });
                    this.addMissingReferences(change.uri, missingReferences);
                    return change.target;
                }
                logger.warn(`Change type "${change.type}" is not supported when "writeSapAnnotations" parameter is set.`);
                return undefined;
            })
                .filter((target) => !!target)
        });
        for (const target of targets) {
            writer.addChange(createInsertTargetChange('target', target));
        }
    }
    getWriterForChange(writers, change) {
        const cachedWriter = writers.get(change.uri);
        if (cachedWriter) {
            return cachedWriter;
        }
        const writer = this.createWriter(change);
        writers.set(change.uri, writer);
        return writer;
    }
    createEmptyAnnotationFile(change) {
        return {
            type: ANNOTATION_FILE_TYPE,
            uri: change.uri,
            references: [],
            targets: []
        };
    }
    createWriter(change) {
        const document = this.documents.get(change.uri);
        if (!document && this.ignoreChangedFileInitialContent) {
            if (!this.facade) {
                throw new Error(`CDS facade not available!`);
            }
            if (change.type !== INSERT_TARGET_CHANGE_TYPE) {
                throw new Error(`Change "${change.type}" type is not supported with parameter "${this.ignoreChangedFileInitialContent}".`);
            }
            const textDocument = TextDocument.create(change.uri, 'cds', 0, '');
            return new CDSWriter(this.facade, this.vocabularyService, {
                type: CDS_DOCUMENT_TYPE,
                uri: change.uri,
                references: [],
                targets: [],
                range: Range.create(0, 0, 0, 0)
            }, [], [], textDocument, this.project.root, this.createEmptyAnnotationFile(change));
        }
        else {
            const cachedFile = this.fileCache?.get(change.uri);
            if (!document || cachedFile === undefined || !this.facade) {
                throw new Error(`CDS document "${change.uri}" is not found!`);
            }
            const textDocument = TextDocument.create(change.uri, 'cds', 0, cachedFile);
            return new CDSWriter(this.facade, this.vocabularyService, document.ast, document.comments, document.tokens, textDocument, this.project.root, document.annotationFile);
        }
    }
    /**
     * Checks if there are no compile errors in the files after update.
     *
     * @param fileCache - Updated file content.
     * @returns Errors.
     */
    async validateChanges(fileCache) {
        return this.sync(fileCache);
    }
    /**
     * Converts annotation object to a string.
     *
     * @param target - Content of an 'Annotations' element.
     * @returns CDS representation of the annotations.
     */
    serializeTarget(target) {
        return printTarget(target);
    }
    _getCompiledService() {
        const annotationFiles = [];
        for (const file of this._fileSequence ?? []) {
            const document = this.documents.get(file.uri);
            const ghostDocument = this.documents.get('!' + file.uri);
            if (ghostDocument) {
                annotationFiles.push(ghostDocument.annotationFile);
            }
            if (document) {
                annotationFiles.push(document.annotationFile);
            }
            else {
                throw new ApiError(`Could not compile service. Missing document ${file.uri}`, ApiErrorCode.General);
            }
        }
        annotationFiles.reverse();
        return Object.freeze({
            odataVersion: '4.0',
            annotationFiles,
            metadata: this.metadata
        });
    }
    clearState() {
        this.missingReferences = {};
    }
    invalidateCaches() {
        this._compiledService = undefined;
        this._fileSequence = undefined;
    }
    updateFileSequence(facade) {
        this._fileSequence = facade.getFileSequence().map((uri) => ({
            uri: pathToFileURL(uri).toString(),
            isReadOnly: uri.indexOf('node_modules') !== -1
        }));
        this.service.serviceFiles = [...this._fileSequence];
    }
    processMissingReferences(uri, writer) {
        const missingReferences = this.missingReferences[uri];
        if (missingReferences?.size) {
            const pointer = '/references';
            const normalizedReferences = [...missingReferences].map((reference) => {
                if (reference.startsWith('file://')) {
                    // uri => convert to relative path to root
                    const path = pathFromUri(reference);
                    return relative(this.project.root, path);
                }
                else {
                    return reference;
                }
            });
            writer.addChange(createInsertReferenceChange(pointer, normalizedReferences));
        }
    }
    addMissingReferences(uri, references) {
        const missingReferences = (this.missingReferences[uri] ??= new Set());
        for (const reference of references) {
            missingReferences.add(reference);
        }
    }
    [INSERT_TARGET] = (writer, document, change) => {
        const aliasInfo = getAliasInfo(document.annotationFile, this.metadataService, this.vocabularyService);
        const missingReferences = getMissingRefs(this.documents, change.uri, change.target.name, change.target, aliasInfo, this.metadataService, {
            apps: Object.keys(this.project.apps),
            projectRoot: this.project.root,
            appName: this.appName
        });
        this.addMissingReferences(document.uri, missingReferences);
        // This has to happen after getting refs, because currently it depends on OData path syntax
        const targetName = change.target.name;
        const pathSegments = targetName.split('/');
        const pathBase = pathSegments.shift() ?? '';
        const parsedName = parseIdentifier(pathBase);
        const fullyQualifiedPath = toFullyQualifiedName(aliasInfo.aliasMap, aliasInfo.currentFileNamespace, parsedName) ?? '';
        const metadataElement = this.metadataService.getMetadataElement(fullyQualifiedPath);
        const fqName = metadataElement?.originalName ?? pathBase;
        let originalPathBase = fqName;
        if (parsedName.namespaceOrAlias !== undefined) {
            const namespace = aliasInfo.aliasMap[parsedName.namespaceOrAlias];
            if (namespace) {
                originalPathBase = originalPathBase.replace(namespace, parsedName.namespaceOrAlias);
            }
        }
        change.target.name = [originalPathBase, ...pathSegments].join('/');
        let complexTypePathSegments;
        if (pathSegments?.[0]?.includes('_')) {
            const segments = pathSegments[0].split('_');
            const complexType = this.metadataService.getMetadataElement(`${fqName}/${segments[0]}`);
            if (complexType?.structuredType && complexType.isComplexType) {
                const element = this.metadataService.getMetadataElement(`${fqName}/${pathSegments[0]}`);
                if (element?.kind === ELEMENT_TYPE) {
                    complexTypePathSegments = pathSegments[0].split('_');
                }
            }
        }
        writer.addChange(createInsertTargetChange('target', change.target, undefined, complexTypePathSegments));
    };
    [DELETE_ELEMENT] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const [currentAstNode, parentAstNode, , greatGrandParentAstNode] = getAstNodesFromPointer(document.ast, pointer).reverse();
        if (currentAstNode?.type === RECORD_PROPERTY_TYPE) {
            writer.addChange({
                type: 'delete-record-property',
                pointer: pointer
            });
        }
        else if (currentAstNode?.type === ANNOTATION_TYPE && parentAstNode.type === RECORD_TYPE) {
            // embedded annotation
            writer.addChange({
                type: 'delete-embedded-annotation',
                pointer: pointer
            });
        }
        else if (currentAstNode.type === TARGET_TYPE) {
            writer.addChange({
                type: 'delete-target',
                pointer: pointer
            });
        }
        else if ((currentAstNode?.type === ANNOTATION_TYPE && parentAstNode.type === TARGET_TYPE) ||
            (currentAstNode?.type === ANNOTATION_TYPE && parentAstNode.type === ANNOTATION_GROUP_ITEMS_TYPE)) {
            writer.addChange({
                type: 'delete-annotation',
                pointer: pointer,
                target: change.target
            });
            checkAndDeleteFlattenedStructures(writer, document, change, parentAstNode, greatGrandParentAstNode);
        }
        else if (currentAstNode?.type === RECORD_TYPE) {
            writer.addChange({
                type: 'delete-record',
                pointer: pointer
            });
        }
        else if (currentAstNode.type === 'boolean' ||
            currentAstNode.type === 'string' ||
            currentAstNode.type === 'path' ||
            currentAstNode.type === 'enum' ||
            currentAstNode.type === 'token' ||
            currentAstNode.type === 'number') {
            writer.addChange({
                type: 'delete-primitive-value',
                pointer: pointer
            });
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} `, ApiErrorCode.General);
        }
    };
    [INSERT_ELEMENT] = (writer, document, change) => {
        const { pointer, containsFlattenedNodes } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const [currentAstNode] = getAstNodesFromPointer(document.ast, pointer).slice(-1);
        if (containsFlattenedNodes) {
            this.insertInFlattenedStructure(writer, document, change, pointer);
        }
        else if (currentAstNode.type === TARGET_TYPE) {
            writer.addChange({
                type: 'insert-annotation',
                pointer: pointer,
                element: change.element
            });
        }
        else if (currentAstNode.type === ANNOTATION_TYPE) {
            this.insertAnnotation(writer, change, pointer);
        }
        else if (currentAstNode.type === RECORD_TYPE) {
            this.insertRecord(writer, change, pointer, currentAstNode);
        }
        else if (currentAstNode.type === RECORD_PROPERTY_TYPE) {
            if (PRIMITIVE_TYPE_NAMES.includes(change.element.name)) {
                writer.addChange({
                    type: INSERT_PRIMITIVE_VALUE_TYPE,
                    pointer: pointer, // point to properties
                    element: change.element,
                    index: change.index
                });
            }
        }
        else if (currentAstNode.type === COLLECTION_TYPE) {
            if (change.element.name === Edm.Record) {
                writer.addChange({
                    type: 'insert-record',
                    pointer: pointer,
                    element: change.element,
                    index: change.index
                });
            }
            else if (PRIMITIVE_TYPE_NAMES.includes(change.element.name)) {
                writer.addChange({
                    type: INSERT_PRIMITIVE_VALUE_TYPE,
                    pointer: pointer,
                    element: change.element,
                    index: change.index
                });
            }
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} ${change.element.name}`, ApiErrorCode.General);
        }
        const aliasInfo = getAliasInfo(document.annotationFile, this.metadataService, this.vocabularyService);
        const missingReferences = getMissingRefs(this.documents, change.uri, change.target, change.element, aliasInfo, this.metadataService, { apps: Object.keys(this.project.apps), projectRoot: this.project.root, appName: '' });
        this.addMissingReferences(document.uri, missingReferences);
    };
    insertInFlattenedStructure(writer, document, change, pointer) {
        const targetPointer = pointer.split('/').slice(0, 3).join('/');
        const termPointer = change.pointer.split('/').slice(0, 5).join('/');
        const annotationValuePointer = change.pointer.split('/').slice(5).join('/');
        const element = getGenericNodeFromPointer(document.annotationFile, termPointer);
        if (element?.type === ELEMENT_TYPE) {
            const annotation = buildAnnotation(element, annotationValuePointer, change.element);
            if (annotation) {
                writer.addChange({
                    type: 'insert-annotation',
                    element: annotation,
                    pointer: targetPointer
                });
            }
        }
    }
    insertAnnotation(writer, change, pointer) {
        // insert annotation value
        if (change.element.name === Edm.Annotation) {
            writer.addChange(createInsertEmbeddedAnnotationChange(pointer, change.element));
        }
        else if (change.element.name === Edm.Collection) {
            writer.addChange(createInsertCollectionChange(pointer, change.element));
        }
        else if (change.element.name === Edm.Record) {
            writer.addChange(createInsertRecordChange(pointer, change.element));
        }
        else if (PRIMITIVE_TYPE_NAMES.includes(change.element.name)) {
            writer.addChange(createInsertPrimitiveValueChange(pointer, change.element));
        }
    }
    insertRecord(writer, change, pointer, record) {
        if (change.element.name === Edm.PropertyValue) {
            const index = adaptRecordPropertyIndex(record, change.index);
            const modifiedPointer = [...pointer.split('/'), 'properties'].join('/'); // pointer is record
            writer.addChange(createInsertRecordPropertyChange(modifiedPointer, change.element, index));
        }
        else if (change.element.name === Edm.Annotation) {
            const index = adaptRecordPropertyIndex(record, change.index);
            writer.addChange(createInsertEmbeddedAnnotationChange(pointer, change.element, index));
        }
        else if (change.element.name === Edm.Record) {
            const segment = pointer.split('/');
            const changeIndex = Number.parseInt(segment.pop() ?? '', 10);
            const modifiedPointer = segment.join('/'); //point to annotations
            writer.addChange(createInsertRecordChange(modifiedPointer, change.element, changeIndex));
        }
    }
    [INSERT_ATTRIBUTE] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const [currentAstNode] = getAstNodesFromPointer(document.ast, pointer).slice(-1);
        if (pointer) {
            if (currentAstNode.type === ANNOTATION_TYPE && change.name === Edm.Qualifier) {
                writer.addChange({
                    type: 'insert-qualifier',
                    pointer: pointer,
                    value: change.value
                });
            }
            if (currentAstNode.type === RECORD_PROPERTY_TYPE ||
                (currentAstNode.type === ANNOTATION_TYPE && change.name !== Edm.Qualifier)) {
                writer.addChange({
                    type: 'insert-primitive-value',
                    pointer: pointer,
                    element: createElementNode({
                        name: change.name,
                        content: [createTextNode(change.value)]
                    })
                });
            }
            if (currentAstNode.type === RECORD_TYPE && change.name === Edm.Type) {
                writer.addChange({
                    type: 'insert-record-property',
                    pointer: pointer + '/properties',
                    index: 0,
                    element: createElementNode({
                        name: 'PropertyValue',
                        attributes: {
                            Property: createAttributeNode('Property', '$Type')
                        },
                        content: [createElementNode({ name: 'String', content: [createTextNode(change.value)] })]
                    })
                });
            }
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} ${change.name}`, ApiErrorCode.General);
        }
    };
    [DELETE_ATTRIBUTE] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const [node] = getAstNodesFromPointer(document.ast, pointer).slice(-1);
        if (pointer && node?.type === QUALIFIER_TYPE) {
            writer.addChange(createDeleteQualifierChange(pointer));
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer}`, ApiErrorCode.General);
        }
    };
    [UPDATE_ATTRIBUTE_VALUE] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        if (pointer) {
            writer.addChange(createUpdatePrimitiveValueChange(pointer, change.newValue));
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} ${change.newValue}`, ApiErrorCode.General);
        }
    };
    [REPLACE_ATTRIBUTE] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const [currentAstNode] = getAstNodesFromPointer(document.ast, pointer).slice(-1);
        if (pointer) {
            if (currentAstNode.type === RECORD_PROPERTY_TYPE && change.newAttributeValue) {
                writer.addChange({
                    type: 'update-primitive-value',
                    pointer: pointer,
                    newValue: change.newAttributeValue
                });
            }
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer}`, ApiErrorCode.General);
        }
    };
    [REPLACE_ELEMENT] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const currentAFNode = getGenericNodeFromPointer(document.annotationFile, change.pointer);
        if (pointer) {
            if (currentAFNode?.type === ELEMENT_TYPE && currentAFNode.name === Edm.PropertyValue) {
                writer.addChange({
                    type: 'replace-record-property',
                    pointer: pointer,
                    newProperty: change.newElement
                });
            }
            else {
                writer.addChange({
                    type: 'replace-node',
                    pointer: pointer,
                    newElement: change.newElement
                });
            }
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} ${change.newElement.name}`, ApiErrorCode.General);
        }
    };
    [REPLACE_TEXT] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const segments = change.pointer.split('/');
        const lastSegment = segments.pop();
        const annotationFileNode = getGenericNodeFromPointer(document.annotationFile, segments.join('/'));
        if (pointer) {
            if (lastSegment === 'text' && elementHasFlags(annotationFileNode)) {
                // CDS has specific syntax for flags and
                writer.addChange({
                    type: 'set-flags',
                    pointer: pointer,
                    value: change.text.text
                });
            }
            else {
                writer?.addChange({
                    type: 'replace-text-value',
                    pointer: pointer,
                    newValue: change.text.text
                });
            }
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} ${change.text}`, ApiErrorCode.General);
        }
    };
    [REPLACE_ELEMENT_CONTENT] = (writer, document, change) => {
        const { pointer } = convertPointer(document.annotationFile, change.pointer, document.ast);
        const [currentAstNode] = getAstNodesFromPointer(document.ast, pointer).slice(-1);
        const annotationFileNode = getGenericNodeFromPointer(document.annotationFile, change.pointer);
        const newValue = change.newValue[0];
        if (pointer && currentAstNode && newValue?.type === TEXT_TYPE) {
            if (annotationFileNode?.type === ELEMENT_TYPE && annotationFileNode.name === Edm.EnumMember) {
                if (elementHasFlags(annotationFileNode)) {
                    writer.addChange({
                        type: 'set-flags',
                        pointer: pointer,
                        value: newValue.text
                    });
                }
                else {
                    writer.addChange({
                        type: 'replace-text-value',
                        pointer: pointer,
                        newValue: newValue.text
                    });
                }
            }
            else {
                const text = 
                // annotation paths are currently converted to strings and for those separators do not need to be replaced
                annotationFileNode?.type === ELEMENT_TYPE &&
                    (annotationFileNode.name === Edm.Path ||
                        annotationFileNode.name === Edm.PropertyPath ||
                        annotationFileNode.name === Edm.NavigationPropertyPath)
                    ? newValue.text.replaceAll('/', '.')
                    : newValue.text;
                writer.addChange({
                    type: 'update-primitive-value',
                    pointer: pointer,
                    newValue: text
                });
            }
        }
        else {
            throw new ApiError(`Could not process change ${change.type} ${change.uri} ${change.pointer} ${change.newValue}`, ApiErrorCode.General);
        }
    };
    [MOVE_ELEMENT] = (writer, document, change) => {
        const { pointer, index } = change;
        const { pointer: toPointer } = convertPointer(document.annotationFile, pointer, document.ast);
        const toNode = getGenericNodeFromPointer(document.annotationFile, pointer);
        if (toPointer && isElementWithName(toNode, 'Collection')) {
            writer.addChange({
                type: 'move-collection-value',
                pointer: toPointer,
                index,
                fromPointers: change.fromPointers.map((ptr) => {
                    const { pointer } = convertPointer(document.annotationFile, ptr, document.ast);
                    return pointer;
                })
            });
        }
    };
    [UPDATE_ELEMENT_NAME] = () => {
        // noop, such changes have no effect in CDS
    };
    [DELETE_REFERENCE] = () => {
        // noop, such changes are not supported in CDS
    };
}
function getAliasInfo(annotationFileInternal, metadataService, vocabularyAPI) {
    const namespaces = getAllNamespacesAndReferences(annotationFileInternal.namespace, annotationFileInternal.references);
    const aliasInfo = getAliasInformation(namespaces, metadataService.getNamespaces());
    return addAllVocabulariesToAliasInformation(aliasInfo, vocabularyAPI.getVocabularies());
}
function elementHasFlags(element) {
    if (element?.type !== 'element') {
        return false;
    }
    const content = element.content[0];
    if (content?.type === 'text' && element.name === Edm.EnumMember) {
        return content.text.split(' ').length > 1;
    }
    return false;
}
function buildAnnotation(root, pointer, lastContent) {
    const segments = pointer.split('/');
    let node = root;
    if (node.type !== ELEMENT_TYPE) {
        return undefined;
    }
    const result = buildElement(node);
    let current = result;
    for (const segment of segments) {
        const next = node[segment];
        if (next) {
            node = next;
            if (node.type === ELEMENT_TYPE) {
                const nextElement = buildElement(node);
                current.content.push(nextElement);
                current = nextElement;
            }
            else if (!Array.isArray(node)) {
                return undefined;
            }
        }
        else {
            return undefined;
        }
    }
    current.content.push(lastContent);
    return result;
}
function buildElement(node) {
    const result = createElementNode({ name: node.name });
    if (node.name === Edm.Annotation) {
        result.attributes[Edm.Term] = node.attributes[Edm.Term];
        if (node.attributes[Edm.Qualifier]) {
            result.attributes[Edm.Qualifier] = node.attributes[Edm.Qualifier];
        }
    }
    if (node.name === Edm.PropertyValue) {
        result.attributes[Edm.Property] = node.attributes[Edm.Property];
    }
    return result;
}
function adaptRecordPropertyIndex(record, currentIndex) {
    if (currentIndex === undefined) {
        return currentIndex;
    }
    let adaptedIdx = currentIndex;
    for (let index = 0; index < record.properties.length; index++) {
        const propertyName = record.properties[index].name.value;
        if (isReservedProperty(propertyName) && index <= adaptedIdx) {
            adaptedIdx = adaptedIdx + 1;
        }
    }
    return adaptedIdx;
}
function checkAndDeleteFlattenedStructures(writer, document, change, parentAstNode, greatGrandParentAstNode) {
    const element = getGenericNodeFromPointer(document.annotationFile, change.pointer);
    if (element?.type === ELEMENT_TYPE) {
        let targetNode;
        if (parentAstNode.type === TARGET_TYPE) {
            targetNode = parentAstNode;
        }
        else if (greatGrandParentAstNode.type === TARGET_TYPE) {
            targetNode = greatGrandParentAstNode;
        }
        if (targetNode) {
            const termName = getElementAttributeValue(element, Edm.Term);
            const qualifier = getElementAttribute(element, Edm.Qualifier);
            if (termName) {
                deleteChildFlattenedStructures(targetNode.name, termName, qualifier?.value, document.ast, writer);
            }
        }
    }
}
// Splitting the logic into multiple functions would make the code more difficult to follow than it currently is.
// eslint-disable-next-line sonarjs/cognitive-complexity
function deleteChildFlattenedStructures(targetName, term, qualifier, ast, writer) {
    for (let targetIndex = 0; targetIndex < ast.targets.length; targetIndex++) {
        const target = ast.targets[targetIndex];
        if (target.name !== targetName) {
            continue;
        }
        for (let assignmentIndex = 0; assignmentIndex < target.assignments.length; assignmentIndex++) {
            const assignment = target.assignments[assignmentIndex];
            if (isMatchingAnnotation(assignment, term, undefined, qualifier)) {
                writer.addChange(createDeleteAnnotationChange(`/targets/${targetIndex}/assignments/${assignmentIndex}`, targetName));
            }
            else if (assignment.type === ANNOTATION_GROUP_TYPE) {
                for (let groupItemIndex = 0; groupItemIndex < assignment.items.items.length; groupItemIndex++) {
                    const item = assignment.items.items[groupItemIndex];
                    const combinedTerm = `${assignment.name.value}.${item.term.value}`;
                    if (isMatchingAnnotation(item, term, combinedTerm, qualifier)) {
                        writer.addChange(createDeleteAnnotationChange(`/targets/${targetIndex}/assignments/${assignmentIndex}/items/items/${groupItemIndex}`, targetName));
                    }
                }
            }
        }
    }
}
function isMatchingAnnotation(node, prefix, term, qualifier) {
    if (node.type !== ANNOTATION_TYPE) {
        return false;
    }
    term ??= node.term.value;
    const [match, next] = term.split(prefix);
    return match === '' && next.startsWith('.') && node.qualifier?.value === qualifier;
}
