// @sap-ux/fiori-annotation-api/dist/cds/change.js
//#region Insert Changes
export const INSERT_TARGET_CHANGE_TYPE = 'insert-target';
/**
 * Creates a change to insert a new annotation target into a document.
 *
 * @param pointer - Pointer to an element.
 * @param target - Internal representation of the target.
 * @param index - Before which child element the new element should be inserted. If omitted will be added at the end.
 * @param complexTypePathSegments
 * @returns Insert target change.
 */
export function createInsertTargetChange(pointer, target, index, complexTypePathSegments) {
    return {
        type: INSERT_TARGET_CHANGE_TYPE,
        pointer,
        target,
        complexTypePathSegments,
        index
    };
}
function insertElementChangeFactory(type) {
    return function (pointer, element, index) {
        return {
            type,
            pointer,
            element,
            index
        };
    };
}
export const INSERT_RECORD_CHANGE_TYPE = 'insert-record';
export const createInsertRecordChange = insertElementChangeFactory(INSERT_RECORD_CHANGE_TYPE);
export const INSERT_COLLECTION_CHANGE_TYPE = 'insert-collection';
export const createInsertCollectionChange = insertElementChangeFactory(INSERT_COLLECTION_CHANGE_TYPE);
export const INSERT_ANNOTATION_CHANGE_TYPE = 'insert-annotation';
export const createInsertAnnotationChange = insertElementChangeFactory(INSERT_ANNOTATION_CHANGE_TYPE);
export const INSERT_EMBEDDED_ANNOTATION_CHANGE_TYPE = 'insert-embedded-annotation';
export const createInsertEmbeddedAnnotationChange = insertElementChangeFactory(INSERT_EMBEDDED_ANNOTATION_CHANGE_TYPE);
export const INSERT_RECORD_PROPERTY_CHANGE_TYPE = 'insert-record-property';
export const createInsertRecordPropertyChange = insertElementChangeFactory(INSERT_RECORD_PROPERTY_CHANGE_TYPE);
export const INSERT_PRIMITIVE_VALUE_TYPE = 'insert-primitive-value';
export const createInsertPrimitiveValueChange = insertElementChangeFactory(INSERT_PRIMITIVE_VALUE_TYPE);
export const INSERT_QUALIFIER_CHANGE_TYPE = 'insert-qualifier';
/**
 * Creates a change to insert a qualifier to an annotation.
 *
 * @param pointer - Pointer to an annotation.
 * @param value - Qualifier.
 * @returns Insert qualifier change.
 */
export function createInsertQualifierChange(pointer, value) {
    return {
        type: INSERT_QUALIFIER_CHANGE_TYPE,
        pointer,
        value
    };
}
export const INSERT_REFERENCE_CHANGE_TYPE = 'insert-reference';
/**
 * Creates a change to insert file references into a document.
 *
 * @param pointer - Pointer to a document.
 * @param references - File references that needs to be added.
 * @returns Insert reference change.
 */
export function createInsertReferenceChange(pointer, references) {
    return {
        type: INSERT_REFERENCE_CHANGE_TYPE,
        pointer,
        references
    };
}
function deleteChangeFactory(type) {
    return function (pointer) {
        return {
            type,
            pointer
        };
    };
}
export const DELETE_TARGET_CHANGE_TYPE = 'delete-target';
export const createDeleteTargetChange = deleteChangeFactory(DELETE_TARGET_CHANGE_TYPE);
export const DELETE_RECORD_CHANGE_TYPE = 'delete-record';
export const createDeleteRecordChange = deleteChangeFactory(DELETE_RECORD_CHANGE_TYPE);
export const DELETE_RECORD_PROPERTY_CHANGE_TYPE = 'delete-record-property';
export const createDeleteRecordPropertyChange = deleteChangeFactory(DELETE_RECORD_PROPERTY_CHANGE_TYPE);
export const DELETE_ANNOTATION_GROUP_CHANGE_TYPE = 'delete-annotation-group';
export const DELETE_ANNOTATION_GROUP_ITEMS_CHANGE_TYPE = 'delete-annotation-group-items';
/**
 * Creates a change to delete an annotation group.
 *
 * @param pointer - pointer to an annotation group
 * @returns Delete annotation group change.
 */
export function createDeleteAnnotationGroupChange(pointer) {
    return {
        type: DELETE_ANNOTATION_GROUP_CHANGE_TYPE,
        pointer
    };
}
/**
 * Creates a change to delete all items within an annotation group.
 *
 * @param pointer - pointer to an annotation group
 * @returns Delete annotation group change.
 */
export function createDeleteAnnotationGroupItemsChange(pointer) {
    return {
        type: DELETE_ANNOTATION_GROUP_ITEMS_CHANGE_TYPE,
        pointer
    };
}
export const DELETE_ANNOTATION_CHANGE_TYPE = 'delete-annotation';
/**
 * Creates a change to delete an annotation.
 *
 * @param pointer - Pointer to an annotation.
 * @param target - Target name.
 * @returns Delete annotation change.
 */
export function createDeleteAnnotationChange(pointer, target) {
    return {
        type: DELETE_ANNOTATION_CHANGE_TYPE,
        pointer,
        target
    };
}
export const DELETE_EMBEDDED_ANNOTATION_CHANGE_TYPE = 'delete-embedded-annotation';
export const createDeleteEmbeddedChange = deleteChangeFactory(DELETE_EMBEDDED_ANNOTATION_CHANGE_TYPE);
export const DELETE_PRIMITIVE_VALUE_CHANGE_TYPE = 'delete-primitive-value';
export const createDeletePrimitiveValueChange = deleteChangeFactory(DELETE_PRIMITIVE_VALUE_CHANGE_TYPE);
export const DELETE_QUALIFIER_CHANGE_TYPE = 'delete-qualifier';
export const createDeleteQualifierChange = deleteChangeFactory(DELETE_QUALIFIER_CHANGE_TYPE);
//#endregion
//#region Modification Changes
export const REPLACE_NODE_CHANGE_TYPE = 'replace-node';
/**
 * Creates a change to replace a node with a new element.
 *
 * @param pointer - Pointer to a node.
 * @param element - Replacement node in internal representation.
 * @returns Replace node change.
 */
export function createReplaceNodeChange(pointer, element) {
    return {
        type: REPLACE_NODE_CHANGE_TYPE,
        pointer,
        newElement: element
    };
}
export const REPLACE_RECORD_PROPERTY_CHANGE_TYPE = 'replace-record-property';
/**
 * Creates a change to replace a record property with a new element.
 *
 * @param pointer - Pointer to a record property.
 * @param element - New Record property in internal representation.
 * @returns Replace record property change.
 */
export function createReplaceRecordPropertyChange(pointer, element) {
    return {
        type: REPLACE_RECORD_PROPERTY_CHANGE_TYPE,
        pointer,
        newProperty: element
    };
}
export const REPLACE_TEXT_VALUE_CHANGE_TYPE = 'replace-text-value';
/**
 * Creates a change to replace a text value with a new text value.
 *
 * @param pointer - Pointer to a primitive value.
 * @param newValue - New text value.
 * @returns Replace text value change.
 */
export function createReplaceTextValueChange(pointer, newValue) {
    return {
        type: REPLACE_TEXT_VALUE_CHANGE_TYPE,
        pointer,
        newValue
    };
}
export const UPDATE_PRIMITIVE_VALUE_CHANGE_TYPE = 'update-primitive-value';
/**
 * Creates a change to update a primitive value.
 *
 * @param pointer - Pointer to a primitive value.
 * @param newValue - New value.
 * @returns Replace primitive value change.
 */
export function createUpdatePrimitiveValueChange(pointer, newValue) {
    return {
        type: UPDATE_PRIMITIVE_VALUE_CHANGE_TYPE,
        pointer,
        newValue
    };
}
//#endregion
export const SET_FLAGS_CHANGE_TYPE = 'set-flags';
/**
 * Creates a change to set enum flags on a collection.
 *
 * @param pointer - Pointer to a enum flags node (collection).
 * @param value - A list of enum member names separated by spaces. Can also contain type information in OData format e.g Communication.PhoneType/work.
 * @returns Set flags change.
 */
export function createSetFlagsChange(pointer, value) {
    return {
        type: SET_FLAGS_CHANGE_TYPE,
        pointer,
        value
    };
}
export const MOVE_COLLECTION_VALUE_CHANGE_TYPE = 'move-collection-value';
/**
 * Creates a change to move nodes within a collection to a new position.
 *
 * @param pointer - Pointer to a container where the nodes will be moved.
 * @param fromPointers - Pointers to nodes which needs to be moved.
 * @param index - Before which child element the new element should be inserted. If omitted will be added at the end.
 * @returns Move collection value change.
 */
export function createMoveCollectionChange(pointer, fromPointers, index) {
    return {
        type: MOVE_COLLECTION_VALUE_CHANGE_TYPE,
        pointer,
        fromPointers,
        index
    };
}
export const CONVERT_TO_COMPOUND_ANNOTATION_CHANGE_TYPE = 'convert-to-compound-annotation';
/**
 * Creates a change to convert an annotation to compound format.
 *
 * @param pointer - Pointer to an assignment.
 * @param applyContentIndentation - Flag indicating if content should be also indented.
 * @returns Convert to compound annotation change.
 */
export function createConvertToCompoundAnnotationChange(pointer, applyContentIndentation) {
    return {
        type: CONVERT_TO_COMPOUND_ANNOTATION_CHANGE_TYPE,
        pointer,
        applyContentIndentation
    };
}
