// @sap-ux/fiori-annotation-api/dist/cds/document.js
import { fileURLToPath } from 'node:url';
import { Range } from '@sap-ux/odata-annotation-core-types';
import { toAnnotationFile, toTargetMap } from '@sap-ux/cds-odata-annotation-converter';
import { compareByRange, pathFromUri } from '../utils/index.js';
import { collectComments } from './comments.js';
export const CDS_DOCUMENT_TYPE = 'document';
/**
 * Creates CDS document.
 *
 * @param serviceName - Name of the service.
 * @param vocabularyService - Vocabulary API.
 * @param facade - CDS compiler facade instance.
 * @param fileCache - File content cache.
 * @param file - File
 * @param metadataCollector - Metadata collector instance.
 * @returns CDS document.
 */
export function getDocument(serviceName, vocabularyService, facade, fileCache, file, metadataCollector) {
    const comments = getComments(fileCache, file, facade, false);
    const cdsAnnotationFile = toTargetMap(facade.blitzIndex.forUri(file.uri), file.uri, vocabularyService, facade);
    const tokens = facade.getTokensForUri(pathFromUri(file.uri));
    const line = fileCache.get(file.uri)?.split('\n').length ?? 0;
    const character = fileCache.get(file.uri)?.split('\n').pop()?.length ?? 0;
    const range = Range.create(0, 0, line, character);
    const cdsDocument = {
        type: 'document',
        uri: file.uri,
        range,
        namespace: cdsAnnotationFile.namespace,
        references: cdsAnnotationFile.references,
        targets: [...(cdsAnnotationFile.targetMap || [])].map(([, value]) => value)
    };
    const annotationFile = toAnnotationFile(file.uri, vocabularyService, cdsAnnotationFile, metadataCollector).file;
    filterTargets(serviceName, annotationFile);
    return {
        uri: file.uri,
        comments,
        ast: cdsDocument,
        annotationFile: annotationFile,
        tokens
    };
}
function getComments(fileCache, file, facade, ignoreComments = true) {
    const content = fileCache.get(file.uri);
    if (content === undefined || content === null) {
        throw new Error(`File ${file.uri} not found in cache!`);
    }
    const tokenVector = facade.getTokensForUri(fileURLToPath(file.uri));
    return ignoreComments ? [] : collectComments(tokenVector);
}
/**
 * Creates ghost file document.
 *
 * @param serviceName - Name of the service.
 * @param vocabularyService - Vocabulary API.
 * @param facade - CDS compiler facade instance.
 * @param fileCache - File content cache.
 * @param file - File
 * @param metadataCollector - Metadata collector instance.
 * @param propagatedTargetMap - Propagation map.
 * @returns Ghost file document.
 */
export function getGhostFileDocument(serviceName, vocabularyService, facade, fileCache, file, metadataCollector, propagatedTargetMap) {
    const comments = getComments(fileCache, file, facade, false);
    const tokens = facade.getTokensForUri(pathFromUri(file.uri));
    const cdsAnnotationFile = toTargetMap(facade.blitzIndex.forUri(file.uri), file.uri, vocabularyService, facade);
    const annotationFile = toAnnotationFile(file.uri, vocabularyService, cdsAnnotationFile, metadataCollector, undefined, propagatedTargetMap).file;
    filterTargets(serviceName, annotationFile);
    annotationFile.uri = '!' + annotationFile.uri;
    const cdsDocument = {
        type: 'document',
        uri: annotationFile.uri,
        namespace: cdsAnnotationFile.namespace,
        references: cdsAnnotationFile.references,
        targets: [...(cdsAnnotationFile.targetMap || [])].map(([, value]) => value)
    };
    return {
        uri: annotationFile.uri,
        comments,
        tokens,
        ast: cdsDocument,
        annotationFile: annotationFile
    };
}
function filterTargets(serviceName, annotationFile) {
    // only allow targets pointing to current service
    const serviceNamespace = annotationFile.namespace?.name === serviceName ? annotationFile.namespace : undefined;
    const aliasName = serviceNamespace ? serviceNamespace.alias : '';
    annotationFile.targets = annotationFile.targets.filter((target) => target.name.startsWith(serviceName + '.') || (aliasName && target.name.startsWith(aliasName + '.')));
}
/**
 * Returns the number of children in the container node.
 *
 * @param container - Container AST node.
 * @returns Number of children.
 */
export function getChildCount(container) {
    return getItems(container).length;
}
/**
 * Returns child nodes for the given container node.
 *
 * @param container - Container AST node.
 * @returns All child nodes of the container.
 */
export function getItems(container) {
    switch (container.type) {
        case 'target':
            return container.assignments;
        case 'record':
            return container.annotations?.length
                ? [...container.properties, ...container.annotations].sort(compareByRange)
                : container.properties;
        case 'annotation-group-items':
        case 'collection':
            return container.items;
        default:
            return [];
    }
}
