// @sap-ux/odata-annotation-core/dist/annotation-file.js
import { ELEMENT_TYPE, Edm, TEXT_TYPE } from '@sap-ux/odata-annotation-core-types';
export const elements = (predicate, element) => (element.content || []).filter((content) => content.type === ELEMENT_TYPE && (!predicate || predicate(content)));
export const elementsWithName = (name, element) => elements((content) => content.name === name, element);
export const getElementAttribute = (element, name) => element?.attributes?.[name];
export const getElementAttributeValue = (element, name) => {
    const attributeNode = getElementAttribute(element, name);
    return attributeNode?.value ?? '';
};
/**
 * Determines whether the node is of element type with given name.
 *
 * @param node node to check
 * @param name name to check
 * @returns boolean check result
 */
export function isElementWithName(node, name) {
    return node?.type === ELEMENT_TYPE && node.name === name;
}
/**
 * Returns true if text doesn't contain any visible charactes (tabs, spaces, line breaks are ignored).
 *
 * @param text
 * @returns boolean check result
 */
function isEmptyText(text) {
    return (text || '').replace(/\s/g, '').length === 0;
}
/**
 * Get text node content of an element.
 * Elements content is supposed to only contain 'Annotation' tags or single text node.
 *
 * @param element
 * @returns TextNode
 */
export function getSingleTextNode(element) {
    let isInvalid = false;
    let firstTextNode = null;
    (element.content || []).forEach((node) => {
        if (!isInvalid) {
            if (node.type === ELEMENT_TYPE && node.name !== Edm.Annotation) {
                isInvalid = true; // child element which is not 'Annotation'
            }
            else if (node.type === TEXT_TYPE && !isEmptyText(node.text)) {
                if (firstTextNode) {
                    isInvalid = true; // second text node
                }
                else {
                    firstTextNode = node;
                }
            }
        }
    });
    return isInvalid ? null : firstTextNode;
}
