// @sap-ux/fiori-annotation-api/dist/fiori-service.js
var __addDisposableResource = (this && this.__addDisposableResource) || function (env, value, async) {
    if (value !== null && value !== void 0) {
        if (typeof value !== "object" && typeof value !== "function") throw new TypeError("Object expected.");
        var dispose, inner;
        if (async) {
            if (!Symbol.asyncDispose) throw new TypeError("Symbol.asyncDispose is not defined.");
            dispose = value[Symbol.asyncDispose];
        }
        if (dispose === void 0) {
            if (!Symbol.dispose) throw new TypeError("Symbol.dispose is not defined.");
            dispose = value[Symbol.dispose];
            if (async) inner = dispose;
        }
        if (typeof dispose !== "function") throw new TypeError("Object not disposable.");
        if (inner) dispose = function() { try { inner.call(this); } catch (e) { return Promise.reject(e); } };
        env.stack.push({ value: value, dispose: dispose, async: async });
    }
    else if (async) {
        env.stack.push({ async: true });
    }
    return value;
};
var __disposeResources = (this && this.__disposeResources) || (function (SuppressedError) {
    return function (env) {
        function fail(e) {
            env.error = env.hasError ? new SuppressedError(e, env.error, "An error was suppressed during disposal.") : e;
            env.hasError = true;
        }
        var r, s = 0;
        function next() {
            while (r = env.stack.pop()) {
                try {
                    if (!r.async && s === 1) return s = 0, env.stack.push(r), Promise.resolve().then(next);
                    if (r.dispose) {
                        var result = r.dispose.call(r.value);
                        if (r.async) return s |= 2, Promise.resolve(result).then(next, function(e) { fail(e); return next(); });
                    }
                    else s |= 1;
                }
                catch (e) {
                    fail(e);
                }
            }
            if (s === 1) return env.hasError ? Promise.reject(env.error) : Promise.resolve();
            if (env.hasError) throw env.error;
        }
        return next();
    };
})(typeof SuppressedError === "function" ? SuppressedError : function (error, suppressed, message) {
    var e = new Error(message);
    return e.name = "SuppressedError", e.error = error, e.suppressed = suppressed, e;
});
import { TextDocument } from 'vscode-languageserver-textdocument';
import { create as createStore } from 'mem-fs';
import { create as createEditor } from 'mem-fs-editor';
import { DiagnosticSeverity } from '@sap-ux/odata-annotation-core-types';
import { getAliasInformation, getAllNamespacesAndReferences } from '@sap-ux/odata-annotation-core';
import { convertMetadataToAvtSchema, convertAnnotationFile, convertTargetAnnotationsToInternal } from './avt/index.js';
import { XMLAnnotationServiceAdapter, getLocalEDMXService, XML_VOCABULARY_SERVICE } from './xml/index.js';
import { getCDSService, CDSAnnotationServiceAdapter, CDS_VOCABULARY_SERVICE } from './cds/index.js';
import { addAllVocabulariesToAliasInformation } from './vocabularies.js';
import { ApiError, ApiErrorCode } from './error.js';
import { pathFromUri } from './utils/index.js';
import { ChangeConverter } from './change-converter.js';
import { join } from 'node:path';
import { pathToFileURL } from 'node:url';
import { logger } from './logger.js';
function getOptionsWithDefaults(options) {
    return {
        commitOnSave: options.commitOnSave ?? true,
        clearFileResolutionCache: options.clearFileResolutionCache ?? false,
        writeSapAnnotations: options.writeSapAnnotations ?? false,
        ignoreChangedFileInitialContent: options.ignoreChangedFileInitialContent ?? false
    };
}
export const COMPILE_ERROR_MSG = 'Update rejected due to changes leading to compile errors';
/**
 * Service for working with OData annotations in SAP Fiori Elements applications.
 *
 */
export class FioriAnnotationService {
    vocabularyAPI;
    adapter;
    changeConverter;
    fs;
    options;
    project;
    serviceName;
    changes = [];
    fileMergeMaps = {};
    fileCache = new Map();
    isInitialSyncCompleted = false;
    projectInfo = {
        appName: '',
        apps: [],
        projectRoot: ''
    };
    /**
     * For creating new instances use the factory function {@link FioriAnnotationService.createService}.
     *
     * @param vocabularyAPI - Vocabulary API instance.
     * @param adapter - Language specific adapter.
     * @param changeConverter ChangeConverter instance.
     * @param fs - `mem-fs-editor` instance.
     * @param options - API configuration.
     * @param project - Project structure.
     * @param serviceName - Name of the service.
     * @param appName - Name of the application
     */
    constructor(vocabularyAPI, adapter, changeConverter, fs, options, project, serviceName, appName) {
        this.vocabularyAPI = vocabularyAPI;
        this.adapter = adapter;
        this.changeConverter = changeConverter;
        this.fs = fs;
        this.options = options;
        this.project = project;
        this.serviceName = serviceName;
        this.projectInfo = {
            apps: Object.keys(project.apps) || [],
            appName: appName || '',
            projectRoot: project.root
        };
    }
    /**
     * Creates new FioriAnnotationService instance for the given Fiori application.
     *
     * @param this - Constructor
     * @param project - Project structure.
     * @param serviceName - Name of the service used by the app.
     * @param appName - Name of the application.
     * @param fs - Optional `mem-fs-editor` instance.
     * @param options - API configuration.
     * @returns FioriAnnotationService instance.
     */
    static async createService(project, serviceName, appName, fs, options = {}) {
        const vocabularyAPI = project.projectType === 'CAPJava' || project.projectType === 'CAPNodejs'
            ? CDS_VOCABULARY_SERVICE
            : XML_VOCABULARY_SERVICE;
        const finalOptions = getOptionsWithDefaults(options);
        const service = await getService(project, serviceName, appName, fs, finalOptions.clearFileResolutionCache);
        const adapter = createAdapter(project, service, vocabularyAPI, appName, finalOptions.writeSapAnnotations, finalOptions.ignoreChangedFileInitialContent);
        // prepare fs editor if not provided
        let fsEditor;
        if (fs) {
            fsEditor = fs;
        }
        else {
            const store = createStore();
            fsEditor = createEditor(store);
        }
        const changeConverter = new ChangeConverter(serviceName, vocabularyAPI, adapter.metadataService, adapter.splitAnnotationSupport, finalOptions.ignoreChangedFileInitialContent);
        const fioriService = new this(vocabularyAPI, adapter, changeConverter, fsEditor, finalOptions, project, serviceName, appName);
        return fioriService;
    }
    /**
     * Returns a metadata service instance.
     *
     * @returns Metadata service instance.
     */
    getMetadataService() {
        return this.adapter.metadataService;
    }
    /**
     * Returns a vocabulary service instance.
     *
     * @returns Vocabulary service instance.
     */
    getVocabularyAPI() {
        return this.vocabularyAPI;
    }
    /**
     * Returns a list of files which describe the service. The order is from least important to the most important file
     * Annotations defined in the last file would overwrite the ones defined in the previous ones.
     * You can only create changes for files which are not marked as `readOnly`.
     *
     * @param [includeGhostFiles] - If set to true will also include ghost files.
     * @returns Text files.
     */
    getAllFiles(includeGhostFiles = false) {
        return this.adapter.getAllFiles(includeGhostFiles);
    }
    /**
     * Refreshes file content from the file system.
     */
    async sync() {
        this.fileCache.clear();
        const files = await Promise.all(this.adapter.getAllFiles().map(async (file) => {
            const path = pathFromUri(file.uri);
            const content = this.fs.read(path) ?? '';
            return { ...file, content };
        }));
        for (const file of files) {
            this.fileCache.set(file.uri, file.content);
        }
        await this.adapter.sync(this.fileCache);
        this.isInitialSyncCompleted = true;
    }
    /**
     * @returns Returns a map with value list references per external service URI.
     */
    getExternalServices() {
        return this.adapter.getValueListReferences();
    }
    /**
     * Refreshes file content from the file system.
     *
     * @param files
     */
    syncExternalServices(files) {
        for (const [, value] of this.adapter.getValueListReferences().entries()) {
            for (const entry of value) {
                for (const uri of entry.uris) {
                    const resource = files.get(uri);
                    this.adapter.syncExternalService(uri, resource?.data ?? '', resource?.localFilePath ?? '');
                }
            }
        }
    }
    /**
     * Provides initial annotation file content.
     *
     * @param filePath - Path to the newly created file.
     * @returns Annotation file content.
     */
    getInitialFileContent(filePath) {
        return this.adapter.getInitialFileContent?.(this.serviceName, filePath) ?? '';
    }
    /**
     * Reads annotations for a specific service in an application.
     *
     * @returns Service metadata in AVT format.
     */
    getSchema() {
        const rawMetadata = {
            version: this.adapter.metadataService.ODataVersion,
            identification: 'metadataFile',
            schema: convertMetadataToAvtSchema(this.adapter.metadataService),
            references: []
        };
        this.fileMergeMaps = mergeAnnotations(this.adapter.compiledService, rawMetadata, this.serviceName, {
            vocabulary: this.vocabularyAPI
        }, this.adapter.splitAnnotationSupport);
        return rawMetadata;
    }
    /**
     * Reads annotations for a specific external service in an application.
     *
     * @returns Service metadata in AVT format.
     */
    getExternalServiceSchema() {
        return this.adapter.getExternalServices().map(({ compiledService, metadataService, uri, localFileUri }) => {
            const env_1 = { stack: [], error: void 0, hasError: false };
            try {
                const ms = __addDisposableResource(env_1, metadataService.useService(uri), false);
                const rawMetadata = {
                    version: metadataService.ODataVersion,
                    identification: 'metadataFile',
                    schema: convertMetadataToAvtSchema(ms),
                    references: []
                };
                this.fileMergeMaps = mergeAnnotations(compiledService, rawMetadata, '', {
                    vocabulary: this.vocabularyAPI
                }, this.adapter.splitAnnotationSupport);
                return { ...rawMetadata, localFileUri };
            }
            catch (e_1) {
                env_1.error = e_1;
                env_1.hasError = true;
            }
            finally {
                __disposeResources(env_1);
            }
        });
    }
    /**
     * Add change(s) to the stack.
     *
     * @param change - Pending changes.
     */
    edit(change) {
        if (Array.isArray(change)) {
            this.changes.push(...change);
        }
        else {
            this.changes.push(change);
        }
    }
    /**
     * Applies the collected changes to the file system.
     *
     * @param options Save options
     * @returns number of files changed.
     */
    async save(options) {
        if (!this.isInitialSyncCompleted) {
            await this.sync();
        }
        const newContent = {};
        logger.log(`Applying changes:\n${JSON.stringify(this.changes, undefined, 2)}`);
        const annotationFileChanges = this.convertChanges(this.changes);
        const workspaceEdit = await this.adapter.getWorkspaceEdit(annotationFileChanges);
        const changes = workspaceEdit.changes ?? {};
        const fileUris = Object.keys(changes);
        // copy the cache to temp to revert if changes caused cds compiler error
        const temp = new Map(this.fileCache);
        for (const fileUri of fileUris) {
            const path = pathFromUri(fileUri);
            const text = this.fs.read(path) ?? '';
            const newText = applyWorkspaceEdits(fileUri, '', workspaceEdit, text);
            this.fileCache.set(fileUri, newText);
            newContent[path] = newText;
        }
        this.changes = [];
        // Validate implicitly syncs, so we don't need to call it again
        if (typeof this.adapter.validateChanges === 'function') {
            const errors = await this.adapter.validateChanges(this.fileCache);
            if (errors?.size) {
                this.fileCache = new Map(temp);
                const messages = compilerMessagesToErrors(errors);
                throw new ApiError(COMPILE_ERROR_MSG, ApiErrorCode.CompileError, messages);
            }
            else {
                // if no compiler error persist the newly applied text edits
                for (const path of Object.keys(newContent)) {
                    const content = newContent[path];
                    this.saveFile(path, content);
                }
            }
        }
        else if (options?.resyncAfterSave) {
            await this.adapter.sync(this.fileCache);
        }
        if (this.options.commitOnSave) {
            await this.commit();
        }
        this.changes = [];
        this.fileMergeMaps = {};
        return { files: fileUris.length };
    }
    /**
     * Utility function: serialize target with annotations.
     *
     * @param targetAnnotations - Annotations that will be serialized.
     * @param fileUri - use alias information from this file; if none provided, use default aliases for vocabularies, non for metadata
     * @returns Annotations serialized as string for the specified language.
     */
    serializeTarget(targetAnnotations, fileUri) {
        const service = this.adapter.compiledService;
        const annotationFileInternal = service.annotationFiles.find((file) => file.uri === fileUri);
        let aliasInfo = null;
        let namespaces = [];
        if (annotationFileInternal) {
            namespaces = getAllNamespacesAndReferences(annotationFileInternal.namespace, annotationFileInternal.references);
        }
        aliasInfo = getAliasInformation(namespaces, this.getMetadataService().getNamespaces());
        if (!annotationFileInternal) {
            aliasInfo = addAllVocabulariesToAliasInformation(aliasInfo, this.vocabularyAPI.getVocabularies());
        }
        return this.adapter.serializeTarget(convertTargetAnnotationsToInternal(targetAnnotations, aliasInfo));
    }
    /**
     * Converts changes to the internal annotation file change format.
     *
     * @param changes - Changes in AVT format.
     * @returns Changes in internal format.
     */
    convertChanges(changes) {
        const compiledService = this.adapter.compiledService;
        let schema;
        if (this.adapter.splitAnnotationSupport && Object.keys(this.fileMergeMaps).length === 0) {
            // make sure merge maps are filled
            schema = this.getSchema();
        }
        const schemaProvider = () => {
            if (schema) {
                return schema;
            }
            else {
                schema = this.getSchema();
                return schema;
            }
        };
        return this.changeConverter.convert(compiledService, this.fileMergeMaps, schemaProvider, changes);
    }
    saveFile(path, content) {
        this.fs.write(path, content);
    }
    async commit() {
        await new Promise((resolve, reject) => {
            this.fs.commit((error) => {
                if (error instanceof Error) {
                    reject(error);
                }
                else if (typeof error === 'string') {
                    reject(new Error(error));
                }
                else if (error) {
                    reject(new Error('Unknown error.'));
                }
                resolve();
            });
        });
    }
}
function applyWorkspaceEdits(fileUri, languageId, workspaceEdits, content) {
    const document = TextDocument.create(fileUri, languageId, 0, content);
    const fileChanges = workspaceEdits.changes ? (workspaceEdits.changes[fileUri] ?? []) : [];
    return TextDocument.applyEdits(document, fileChanges);
}
function mergeAnnotations(compiledService, rawMetadata, serviceName, options, mergeSplitAnnotations) {
    const fileMergeMaps = {};
    for (const annotationFile of [...compiledService.annotationFiles]) {
        const mergeMap = {};
        fileMergeMaps[annotationFile.uri] = mergeMap;
        const targets = convertAnnotationFile(annotationFile, serviceName, {
            addOrigins: true,
            vocabularyService: options.vocabulary,
            mergeSplitAnnotations: mergeSplitAnnotations,
            mergeMap
        });
        rawMetadata.schema.annotations[annotationFile.uri] = targets;
    }
    return fileMergeMaps;
}
async function getService(project, serviceName, appName, fsEditor, clearCache) {
    if (project.projectType === 'EDMXBackend') {
        return getLocalEDMXService(project, serviceName, appName);
    }
    else if (project.projectType === 'CAPJava' || project.projectType === 'CAPNodejs') {
        const fileCache = new Map();
        if (fsEditor) {
            for (const [relativePath, value] of Object.entries(fsEditor.dump())) {
                const absolute = pathToFileURL(join(process.cwd(), relativePath)).toString();
                fileCache.set(absolute, value.contents);
            }
        }
        return getCDSService(project.root, serviceName, fileCache, clearCache);
    }
    else {
        throw new Error(`Unsupported project type "${project.projectType}"!`);
    }
}
function createAdapter(project, service, vocabularyService, appName, writeSapAnnotations, ignoreChangedFileInitialContent) {
    if (service.type === 'local-edmx') {
        return new XMLAnnotationServiceAdapter(service, vocabularyService, project, appName);
    }
    else if (service.type === 'cap-cds') {
        return new CDSAnnotationServiceAdapter(service, project, vocabularyService, appName, writeSapAnnotations, ignoreChangedFileInitialContent);
    }
    else {
        throw new Error(`Unsupported service type "${service.type}"!`);
    }
}
function compilerMessagesToErrors(compileMessages) {
    const result = new Map();
    [...compileMessages.entries()].forEach((entry) => {
        const [file, value] = entry;
        if (value.hasSyntaxErrors && !file.startsWith('../')) {
            const errorMsgs = value.messages
                .filter((value) => value.severity === DiagnosticSeverity.Error)
                .map((msg) => msg.message);
            result.set(file, errorMsgs);
        }
    });
    return result;
}
