// @sap-ux/cds-odata-annotation-converter/dist/printer/primitives.js
import { Edm } from '@sap-ux/odata-annotation-core';
export const PRIMITIVE_VALUE_ATTRIBUTE_NAMES = new Set([
    Edm.Bool,
    Edm.Decimal,
    Edm.Duration,
    Edm.Float,
    Edm.Int,
    Edm.Null,
    Edm.DateTimeOffset,
    Edm.TimeOfDay,
    Edm.Date,
    Edm.String,
    Edm.EnumMember,
    Edm.Guid,
    Edm.Binary,
    Edm.AnnotationPath,
    Edm.ModelElementPath,
    Edm.NavigationPropertyPath,
    Edm.PropertyPath,
    Edm.Path
]);
export const list = (entries, useTrailingComma = true) => entries.map((entry, i, arr) => {
    const comma = i + 1 !== arr.length || useTrailingComma ? ',' : '';
    if (typeof entry === 'string') {
        return entry + comma;
    }
    else {
        if (entry.placeholder) {
            return entry.value;
        }
        return entry.value + comma;
    }
});
export const container = (open, close, entries) => [open, ...list(entries), close].join('\n');
export const struct = (properties) => container('{', '}', properties);
export const collection = (items) => container('[', ']', items);
export const stringLiteral = (value) => value.indexOf('\n') !== -1 ? `\`\`\`${value}\`\`\`` : `'${value}'`;
export const delimitedIdentifier = (value) => `![${value}]`;
export const valuePair = (key, value) => `${key} : ${value}`;
export const keyAlone = (key) => `${key}`;
export const CDS_NULL_EXPRESSION_LITERAL = 'null';
