// @sap-ux/fiori-annotation-api/dist/avt/expressions.js
/**
 * all attribute or node names that can provide a value representing an Expression (according to typescript definition)
 *  - each entry has the corresponding primitive JS type as string value
 */
export const expressionNames = {
    String: 'string',
    Bool: 'boolean',
    Decimal: 'number',
    Date: 'string',
    Float: 'number',
    Int: 'number',
    Path: 'string',
    PropertyPath: 'string',
    AnnotationPath: 'string',
    NavigationPropertyPath: 'string',
    EnumMember: 'string',
    Collection: 'array',
    Record: 'object',
    Apply: 'object',
    If: 'object',
    And: 'object',
    Or: 'object',
    Le: 'object',
    Lt: 'object',
    Ge: 'object',
    Gt: 'object',
    Eq: 'object',
    Ne: 'object',
    Not: 'object',
    Null: 'null'
};
/**
 * Type guard to check if a value is an annotation expression.
 *
 * @param value - Node.
 * @returns True if node is an expression.
 */
export function isExpression(value) {
    return typeof value === 'object' && value.type !== undefined && expressionNames[value.type] !== undefined;
}
