// @sap-ux/odata-annotation-core-types/dist/annotation-file.js
/**
 * representation of an XML node
 */
export const ATTRIBUTE_TYPE = 'attribute';
export const TEXT_TYPE = 'text';
export var MultilineType;
(function (MultilineType) {
    /**
     * Processor should strip all the indentation before using the value
     */
    MultilineType["StripIndentation"] = "StripIndentation";
    /**
     * Processor should keep the indentation as a part of the value
     */
    MultilineType["KeepIndentation"] = "KeepIndentation";
})(MultilineType || (MultilineType = {}));
export const ELEMENT_TYPE = 'element';
export const EDMX_NAMESPACE_ALIAS = 'Edmx';
export const EDM_NAMESPACE_ALIAS = 'Edm';
export const REFERENCE_TYPE = 'reference';
export const NAMESPACE_TYPE = 'namespace';
export const TARGET_TYPE = 'target';
export const ANNOTATION_FILE_TYPE = 'annotation-file';
/**
 * Creates attribute node.
 *
 * @param name name
 * @param value value
 * @param nameRange name range
 * @param valueRange value range
 * @returns attribute node object
 */
export function createAttributeNode(name, value, nameRange, valueRange) {
    const attribute = {
        type: ATTRIBUTE_TYPE,
        name,
        value
    };
    if (nameRange) {
        attribute.nameRange = nameRange;
    }
    if (valueRange) {
        attribute.valueRange = valueRange;
    }
    return attribute;
}
/**
 * Creates text node.
 *
 * @param text text
 * @param range text range
 * @param fragmentRanges text fragment ranges
 * @param multilineType - multiline type
 * @returns text node object
 */
export function createTextNode(text, range, fragmentRanges, multilineType) {
    const node = { type: TEXT_TYPE, text };
    if (range) {
        node.range = range;
    }
    if (fragmentRanges) {
        node.fragmentRanges = fragmentRanges;
    }
    if (multilineType) {
        node.multilineType = multilineType;
    }
    return node;
}
/**
 * Creates element node.
 *
 * @param param0  object with parameters
 * @param param0.name  name
 * @param param0.range  range
 * @param param0.nameRange  name range
 * @param param0.attributes  attributes
 * @param param0.content  content
 * @param param0.contentRange  content range
 * @param param0.namespace  namespace
 * @param param0.namespaceAlias  alias
 * @returns Element object
 */
export function createElementNode({ name, range, nameRange, attributes, content, contentRange, namespace, namespaceAlias }) {
    const node = {
        type: ELEMENT_TYPE,
        name,
        attributes: attributes ?? {},
        content: content ?? []
    };
    if (range) {
        node.range = range;
    }
    if (nameRange) {
        node.nameRange = nameRange;
    }
    if (contentRange) {
        node.contentRange = contentRange;
    }
    if (namespace) {
        node.namespace = namespace;
    }
    if (namespaceAlias) {
        node.namespaceAlias = namespaceAlias;
    }
    return node;
}
export const GHOST_FILENAME_PREFIX = '!';
/**
 * Builds empty target (generic annotation file format).
 *
 * @param path - target name
 * @returns object - empty target
 */
export function createTarget(path) {
    return { type: 'target', name: path, terms: [], range: undefined, nameRange: undefined, termsRange: undefined };
}
/**
 * Builds empty namespace (generic annotation file format).
 *
 * @param namespace - namespace name
 * @param alias - alias name
 * @param ranges - object with ranges
 * @param ranges.range - element range
 * @param ranges.nameRange - name range
 * @param ranges.aliasRange alias range
 * @param ranges.contentRange - content range
 * @returns Namespace obejct
 */
export function createNamespace(namespace, alias, ranges) {
    const { aliasRange, contentRange, nameRange, range } = ranges ?? {};
    return {
        type: 'namespace',
        name: namespace,
        alias,
        range,
        nameRange,
        aliasRange,
        contentRange
    };
}
/**
 * Builds empty reference (generic annotation file format).
 *
 * @param name reference name
 * @param alias alias
 * @param uri uri
 * @param ranges object with ranges
 * @param ranges.range - element range
 * @param ranges.nameRange - name range
 * @param ranges.aliasRange alias range
 * @param ranges.uriRange - uri range
 * @returns reference onject
 */
export function createReference(name, alias, uri, ranges) {
    const { aliasRange, uriRange, nameRange, range } = ranges ?? {};
    return {
        type: 'reference',
        name,
        alias,
        uri,
        range,
        nameRange,
        aliasRange,
        uriRange
    };
}
