// @sap-ux/odata-annotation-core-types/dist/diagnostics.js
export const NO_UNUSED_NAMESPACE_TYPE = 'no-unused-namespace';
export const NO_UNDEFINED_NAMESPACE_TYPE = 'no-undefined-namespace';
export const NAME_CASE_ISSUE_PATH_VALUE = 'name-case-issue-path-value';
export const MISSING_I18N_KEY = 'missing-i18n-key';
export const VALUE_REQUIRED = 'value-required';
export const NO_WHITESPACE_IN_PATH_EXPRESSION = 'no-whitespace-in-path-expression';
export const INCOMPLETE_EXPRESSION_CC_FORWARD_SLASH = 'incomplete-expression-cc-forward-slash';
export const INCOMPLETE_EXPRESSION_FORWARD_SLASH = 'incomplete-expression-forward-slash';
export const IGNORE_TARGET_VALIDATION = 'ignore-target-validation';
export const UNKNOWN_TERM = 'unknown-term';
export const UN_SUPPORTED_VOCABULARY = 'un-supported-vocabulary';
export const ATTRIBUTE_NOT_ALLOWED_HERE = 'attribute-not-allowed-here';
export const MISSING_REQUIRED_PROPERTY = 'missing-required-property';
export const MISSING_REQUIRED_ATTRIBUTE = 'missing-required-attribute';
export const MISSING_REQUIRED_VALUE_FOR_ATTRIBUTE = 'missing-required-value-for-attribute';
export const TERM_NOT_APPLICABLE = 'term-not-applicable';
export const NOT_IN_APPLICABLE_TERMS_CONSTRAINT = 'not-in-applicable-terms-constraint';
export const RECORD_COLLECTION_PATH_NOT_ALLOWED = 'record-collection-path-not-allowed';
export const ODATA_FUNCTION_WRONG_RETURN_TYPE = 'odata-function-wrong-return-type';
export const IGNORE_DUPLICATE = 'ignore-duplicate';
export const INVALID_PATH_EXPRESSION = 'invlid-path-expression';
export const INVALID_ENUM_MEMBER_TYPE = 'unknown-enum-member';
export const INVALID_TYPE_TYPE = 'invalid-type';
export const NO_VALIDATION_FOR_SUBNODES = 'no-validation-subnodes';
export const INCOMPLETE_PATH_WITH_TYPE = 'incomplete-path-with-type';
export const INCOMPLETE_PATH_WITH_COMPATIBLE_TYPES = 'incomplete-path-with-compatible-types';
export const COMMON_CASE_ISSUE = 'common-case-issue';
export const ODATA_PATH_SEPARATOR_RULE = 'no-odata-path-separator';
export const INVALID_PRIMITIVE_TYPE = 'invalid-primitive-type';
export const DEPRECATED_$VALUE_SYNTAX = 'deprecated-$value-syntax';
