// @sap-ux/logger/dist/transports/transport.js
import WinstonTransport from 'winston-transport';
import { LogLevel, Transport } from '../types.js';
/**
 *  This represents the console target
 */
export class ConsoleTransport extends Transport {
    static singletonInstance;
    options;
    /**
     *
     * @param opts
     */
    constructor(opts = {}) {
        super();
        if (!ConsoleTransport.singletonInstance) {
            ConsoleTransport.singletonInstance = this;
            this.options = this.copy(opts);
        }
        return ConsoleTransport.singletonInstance;
    }
}
/**
 * Transport used in UI5 tooling to print to the console.
 * Don't use this together with `ConsoleTransport` or you'll have logs doubled up
 * on the console in different formats
 */
export class UI5ToolingTransport extends Transport {
    static instances = new Map();
    options;
    /**
     *
     * @param opts
     */
    constructor(opts) {
        super();
        const instance = UI5ToolingTransport.instances.get(opts.moduleName);
        if (!instance) {
            this.options = this.copy(opts);
            UI5ToolingTransport.instances.set(opts.moduleName, this);
            return this;
        }
        else {
            return instance;
        }
    }
}
/**
 *  Use this when you just want a sink for the logs
 */
export class NullTransport extends Transport {
    static singletonInstance;
    /**
     *
     */
    constructor() {
        super();
        if (!NullTransport.singletonInstance) {
            NullTransport.singletonInstance = this;
        }
        return NullTransport.singletonInstance;
    }
}
/**
 *  This represents a file target
 */
export class FileTransport extends Transport {
    options;
    /**
     * Constructor for FileTransport, expects the options for the transport, file name is mandatory.
     *
     * @param opts - options for the transport
     */
    constructor(opts) {
        super();
        this.options = this.copy(opts);
    }
}
/**
 *
 */
export class StringArrayTransport extends Transport {
}
/**
 *  This represents an output channel in VS Code
 *  https://code.visualstudio.com/api/extension-capabilities/common-capabilities#output-channel
 */
export class VSCodeTransport extends Transport {
    static instances = new Map();
    options;
    /**
     *
     * @param opts
     */
    constructor(opts) {
        super();
        const instance = VSCodeTransport.instances.get(opts.channelName);
        if (!instance) {
            this.options = this.copy(opts);
            VSCodeTransport.instances.set(opts.channelName, this);
            return this;
        }
        else {
            return instance;
        }
    }
}
/**
 *
 */
export class ArrayTransport extends WinstonTransport {
    logs;
    /**
     *
     * @param opts
     */
    constructor(opts) {
        super({ level: typeof opts?.logLevel === 'number' ? LogLevel[opts.logLevel].toLowerCase() : 'debug' });
        this.logs = opts?.logs ?? [];
    }
    /**
     *
     * @param info
     * @param next
     */
    log(info, next) {
        this.logs.push(info);
        next();
    }
    // Mixin from Transport
    copy = Transport.prototype.copy;
}
