// @sap-ux/logger/dist/extension-logger/index.js
import { VSCodeTransport } from '../transports/index.js';
import { VSCodeTransport as WinstonVSCodeTransport } from '../winston-logger/vscode-output-channel-transport.js';
import { WinstonLogger } from '../winston-logger/index.js';
import { LogLevel } from '../types.js';
import { toWinstonLogLevel } from '../winston-logger/adapter.js';
/**
 *
 */
export class ExtensionLogger extends WinstonLogger {
    /**
     *
     * @param channelName
     */
    constructor(channelName) {
        const vscodeTransport = new VSCodeTransport({ channelName });
        super({
            logLevel: LogLevel.Silly, // set to lowest level, let VSCode filter levels
            transports: [vscodeTransport],
            logPrefix: ''
        });
    }
    /**
     * Private function to enable adding additional arguments to the log message.
     * Prepares the message for formatting with format.splat().
     *
     * @param level - log level
     * @param message - log message
     * @param args - additional arguments like objects, arrays, etc.
     */
    logWithArgs(level, message, ...args) {
        const winstonLevel = toWinstonLogLevel(level) ?? this._logger.level;
        if (args.length > 0) {
            message += ' %O'.repeat(args.length);
        }
        this._logger.log(winstonLevel, message, ...args);
    }
    /**
     * Log an error message.
     *
     * @param message - error message
     * @param args - additional arguments like objects, arrays, etc.
     */
    error(message, ...args) {
        this.logWithArgs(LogLevel.Error, message, ...args);
    }
    /**
     * Log a warning message.
     *
     * @param message - warning message
     * @param args - additional arguments like objects, arrays, etc.
     */
    warn(message, ...args) {
        this.logWithArgs(LogLevel.Warn, message, ...args);
    }
    /**
     * Log an info message.
     *
     * @param message - info message
     * @param args - additional arguments like objects, arrays, etc.
     */
    info(message, ...args) {
        this.logWithArgs(LogLevel.Info, message, ...args);
    }
    /**
     * Log a debug message.
     *
     * @param message - debug message
     * @param args - additional arguments like objects, arrays, etc.
     */
    debug(message, ...args) {
        this.logWithArgs(LogLevel.Debug, message, ...args);
    }
    /**
     * Log a trace message.
     *
     * @param message - log message
     * @param args - additional arguments like objects, arrays, etc.
     */
    trace(message, ...args) {
        this.logWithArgs(LogLevel.Silly, message, ...args);
    }
    /**
     * Show the output channel in Visual Studio Code.
     */
    show() {
        const winstonVSCodeTransport = this._logger.transports.find((t) => t instanceof WinstonVSCodeTransport);
        if (winstonVSCodeTransport) {
            winstonVSCodeTransport.show();
        }
    }
}
