// @sap-ux/fiori-annotation-api/dist/sap/builders.js
import { Edm, createElementNode, createAttributeNode, Location } from '@sap-ux/odata-annotation-core-types';
function createComplexElement(elementName, nameAttribute, name, content) {
    return createElementNode({
        name: elementName,
        attributes: {
            [nameAttribute]: createAttributeNode(nameAttribute, name)
        },
        content
    });
}
function createPrimitiveElement(elementName, nameAttribute, name, valueType, value) {
    return createElementNode({
        name: elementName,
        attributes: {
            [nameAttribute]: createAttributeNode(nameAttribute, name),
            [valueType]: createAttributeNode(valueType, value)
        }
    });
}
/**
 * Factory function for PropertyValue node with a primitive (attribute) value.
 *
 * @param name - Name of the property.
 * @param valueType - Type of the property.
 * @param value - Value of the property.
 * @returns PropertyValue node.
 */
export function createPrimitiveRecordProperty(name, valueType, value) {
    return createPrimitiveElement(Edm.PropertyValue, Edm.Property, name, valueType, value);
}
/**
 * Factory function for PropertyValue node with a complex (child element) value.
 *
 * @param name - Name of the property.
 * @param value - Value of the property.
 * @returns PropertyValue node.
 */
export function createComplexRecordProperty(name, value) {
    return createComplexElement(Edm.PropertyValue, Edm.Property, name, [value]);
}
/**
 * Factory function for Annotation node with a primitive (attribute) value.
 *
 * @param termName - Annotation term.
 * @param valueType - Type of the annotation value.
 * @param value - Value of the annotation.
 * @returns Annotation node.
 */
export function createPrimitiveAnnotation(termName, valueType, value) {
    return createPrimitiveElement(Edm.Annotation, Edm.Term, termName, valueType, value);
}
/**
 * Factory function for Annotation node with a primitive (child element) value.
 *
 * @param termName - Annotation term.
 * @param value - Value of the annotation.
 * @returns Annotation node.
 */
export function createComplexAnnotation(termName, value) {
    return createElementNode({
        name: Edm.Annotation,
        attributes: {
            [Edm.Term]: createAttributeNode(Edm.Term, termName)
        },
        content: [value]
    });
}
/**
 * Factory function for Record node.
 *
 * @param properties - Record properties.
 * @param type - Record type.
 * @returns Record node.
 */
export function createRecord(properties, type) {
    const node = createElementNode({
        name: Edm.Record,
        content: properties
    });
    if (type !== undefined) {
        node.attributes[Edm.Type] = createAttributeNode(Edm.Type, type);
    }
    return node;
}
/**
 * Factory function for value with origin.
 *
 * @param value - Value.
 * @param uri - File URI where the value is defined.
 * @param range - Range of the value in the file.
 * @returns ValueWithOrigin node.
 */
export function createValue(value, uri, range) {
    const result = {
        value
    };
    if (uri && range) {
        result.location = Location.create(uri, range);
    }
    return result;
}
