// @sap-ux/eslint-plugin-fiori-tools/lib/language/diagnostics.js
export const WIDTH_INCLUDING_COLUMN_HEADER_RULE_TYPE = 'sap-width-including-column-header';
export const ANCHOR_BAR_VISIBLE = 'sap-anchor-bar-visible';
export const FLEX_ENABLED = 'sap-flex-enabled';
export const COPY_TO_CLIPBOARD = 'sap-copy-to-clipboard';
export const ENABLE_EXPORT = 'sap-enable-export';
export const ENABLE_PASTE = 'sap-enable-paste';
export const CREATION_MODE_FOR_TABLE = 'sap-creation-mode-for-table';
export const STATE_PRESERVATION_MODE = 'sap-state-preservation-mode';
export const TABLE_PERSONALIZATION = 'sap-table-personalization';
export const TABLE_COLUMN_VERTICAL_ALIGNMENT = 'sap-table-column-vertical-alignment';
export const TEXT_ARRANGEMENT_HIDDEN = 'sap-text-arrangement-hidden';
export const NO_DATA_FIELD_INTENT_BASED_NAVIGATION = 'sap-no-data-field-intent-based-navigation';
export const CONDENSED_TABLE_LAYOUT = 'sap-condensed-table-layout';
export const STRICT_UOM_FILTERING = 'sap-strict-uom-filtering';
export const DESCRIPTION_COLUMN_LABEL = 'sap-description-column-label';
