// @sap-ux/fiori-annotation-api/dist/xml/changes.js
export const INSERT_ELEMENT = 'insert-element';
export const REPLACE_ELEMENT = 'replace-element';
export const REPLACE_ELEMENT_CONTENT = 'replace-element-content';
export const UPDATE_ELEMENT_NAME = 'update-element-name';
export const DELETE_ELEMENT = 'delete-element';
export const INSERT_ATTRIBUTE = 'insert-attribute';
export const UPDATE_ATTRIBUTE_NAME = 'update-attribute-name';
export const UPDATE_ATTRIBUTE_VALUE = 'update-attribute-value';
export const DELETE_ATTRIBUTE = 'delete-attribute';
export const MOVE_COLLECTION_VALUE = 'move-collection-value';
/**
 *  Creates {@link InsertElement} change.
 *
 * @param pointer - Pointer to the element.
 * @param element - Value to be inserted.
 * @param index - Position before which element should the new value be inserted.
 * @returns InsertElement change object.
 */
export function insertElement(pointer, element, index) {
    const change = {
        type: 'insert-element',
        pointer,
        element
    };
    if (index !== undefined) {
        change.index = index;
    }
    return change;
}
