// @sap-ux/cds-odata-annotation-converter/dist/transforms/annotation/handlers/collection.js
import { COLLECTION_TYPE, nodeRange } from '@sap-ux/cds-annotation-parser';
import { createElementNode, Edm } from '@sap-ux/odata-annotation-core-types';
import { ENUM_TYPE_KIND } from '@sap-ux/odata-vocabularies';
import { convertFlags } from './enum.js';
export const collectionHandler = {
    type: COLLECTION_TYPE,
    getChildren,
    convert(state, node) {
        const valueType = state.context.valueType;
        const vocabularyType = valueType ? state.vocabularyService.getType(valueType) : undefined;
        state.pushContext({ ...state.context, isCollection: false });
        if (valueType && vocabularyType?.kind === ENUM_TYPE_KIND) {
            return convertFlags(state, node, valueType);
        }
        const element = createElementNode({
            name: Edm.Collection,
            range: nodeRange(node, true),
            contentRange: nodeRange(node, false)
        });
        return element;
    }
};
/**
 * Gets the children of a collection node based on the provided visitor state and collection node.
 *
 * @param state - The visitor state.
 * @param node - The collection node.
 * @returns The array of children nodes.
 */
function getChildren(state, node) {
    const valueType = state.context.valueType;
    const vocabularyType = valueType ? state.vocabularyService.getType(valueType) : undefined;
    if (vocabularyType?.kind === ENUM_TYPE_KIND) {
        return [];
    }
    return node.items;
}
