// @sap-ux/cds-odata-annotation-converter/dist/transforms/annotation/handlers/quoted-literal.js
import { QUOTED_LITERAL_TYPE, nodeRange } from '@sap-ux/cds-annotation-parser';
import { createElementNode, Edm } from '@sap-ux/odata-annotation-core-types';
export const quotedLiteralHandler = {
    type: QUOTED_LITERAL_TYPE,
    convert(state, node) {
        const element = createElementNode({
            name: getElementNameForQuotedLiteral(node.kind),
            nameRange: nodeRange(node, false),
            range: nodeRange(node, true)
        });
        return element;
    }
};
/**
 * Converts a quoted literal kind to its corresponding EDM element name.
 *
 * @param kind quoted string.
 * @returns check case and return Edm (Time of Day, date, dateOffset or binary).
 */
function getElementNameForQuotedLiteral(kind) {
    switch (kind) {
        case 'time':
            return Edm.TimeOfDay;
        case 'date':
            return Edm.Date;
        case 'timestamp':
            return Edm.DateTimeOffset;
        case 'binary':
            return Edm.Binary;
        default:
            // Use 'assertNever' to handle exhaustiveness check
            return assertNever(kind);
    }
}
/**
 * Throws an error indicating that a value should never be reached.
 *
 * @param value - The value that should never be reached.
 */
function assertNever(value) {
    throw new Error(`Unexpected value: ${value}`);
}
