// @sap-ux/cds-odata-annotation-converter/dist/transforms/annotation/handlers/record-property.js
import { RECORD_PROPERTY_TYPE, nodeRange } from '@sap-ux/cds-annotation-parser';
import { createElementNode, Edm, Position } from '@sap-ux/odata-annotation-core-types';
import { convertFlattenedPath } from '../flattened.js';
import { getImplicitPropertyType } from '../type-resolver.js';
import { createPropertyAttribute, createTermAttribute } from '../creators.js';
export const recordPropertyHandler = {
    type: RECORD_PROPERTY_TYPE,
    getChildren(state, node) {
        if (node.value) {
            return [node.value];
        }
        return [];
    },
    convert(state, node) {
        if (state.context.inValueContainer) {
            // Since only annotations are allowed here, we assume it is one
            const element = createElementNode({
                name: Edm.Annotation,
                range: nodeRange(node, true)
            });
            element.attributes[Edm.Term] = createTermAttribute(node.name.value, nodeRange(node.name, false));
            if (node.value) {
                element.contentRange = nodeRange(node.value, true);
            }
            return element;
        }
        if (node.name.segments.length === 1) {
            const element = createElementNode({
                name: Edm.PropertyValue,
                range: nodeRange(node, true)
            });
            element.attributes[Edm.Property] = createPropertyAttribute(node.name.value, nodeRange(node.name, false));
            const recordPropertyType = getImplicitPropertyType(state.vocabularyService, state.context, node.name.value)?.type;
            state.pushContext({ ...state.context, valueType: recordPropertyType, propertyName: node.name.value });
            if (node.value) {
                element.contentRange = nodeRange(node.value, true);
                // take into account colon
                if (node.colon && element.contentRange) {
                    const colonPosition = node?.colon?.range?.end;
                    // position right after colon should be counted as part of the elements content
                    element.contentRange.start = colonPosition
                        ? Position.create(colonPosition.line, colonPosition.character)
                        : Position.create(0, 0);
                }
            }
            return element;
        }
        else {
            return convertFlattenedPath(state, node.name.segments, node.value);
        }
    }
};
