// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-localstorage.js
/**
 * @file detects usage of localstaorage
 */
import { createStorageRuleHelpers } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            localStorageUsage: 'Local storage must not be used in a Fiori application'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_STORAGE_OBJECT = [];
        const storageHelpers = createStorageRuleHelpers('localStorage', 'localStorageUsage');
        return {
            'VariableDeclarator'(node) {
                storageHelpers.processVariableDeclarator(node, FORBIDDEN_STORAGE_OBJECT);
            },
            'MemberExpression'(node) {
                storageHelpers.checkMemberExpression(node, FORBIDDEN_STORAGE_OBJECT, context);
            }
        };
    }
};
export default rule;
