// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-element-creation.js
/**
 * @file Detect direct DOM insertion
 */
import { isIdentifier, isCall, isLiteral, contains, createDocumentBasedRuleVisitors } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            elementCreation: 'Direct element creation, create a custom control instead'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_DOM_INSERTION = [
            'createElement',
            'createTextNode',
            'createElementNS',
            'createDocumentFragment',
            'createComment',
            'createAttribute',
            'createEvent'
        ];
        const createVisitors = createDocumentBasedRuleVisitors({
            isInteresting: (node, isDocumentObject) => {
                return node && isCall(node.parent) && isDocumentObject(node.object);
            },
            isValid: (node) => {
                let methodName = false;
                if (isIdentifier(node.property)) {
                    methodName = node.property.name;
                }
                else if (isLiteral(node.property)) {
                    methodName = node.property.value;
                }
                return (methodName &&
                    (!contains(FORBIDDEN_DOM_INSERTION, methodName) ||
                        (methodName === 'createElement' &&
                            isCall(node.parent) &&
                            node.parent.arguments &&
                            node.parent.arguments.length > 0 &&
                            isLiteral(node.parent.arguments[0]) &&
                            node.parent.arguments[0].value === 'a')));
            },
            messageId: 'elementCreation'
        });
        return createVisitors(context);
    }
};
export default rule;
