// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-exec-command.js
/**
 * @file Detect usage of execCommand
 */
import { isIdentifier, isLiteral, createIsWindowObject, createRememberWindow, createIsDocument, createIsDocumentObject, createRememberDocument } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            execCommand: 'Direct DOM Manipulation, better to use jQuery.appendTo if really needed'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const DOCUMENT_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = createIsWindowObject(WINDOW_OBJECTS);
        const rememberWindow = createRememberWindow(WINDOW_OBJECTS, isWindowObject);
        const isDocument = createIsDocument(isWindowObject);
        const isDocumentObject = createIsDocumentObject(DOCUMENT_OBJECTS, isDocument);
        const rememberDocument = createRememberDocument(DOCUMENT_OBJECTS, isDocumentObject);
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberDocument(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right) || rememberDocument(node.left, node.right);
            },
            'MemberExpression': function (node) {
                if (isDocumentObject(node.object) &&
                    ((isIdentifier(node.property) && 'name' in node.property && node.property.name === 'execCommand') ||
                        (isLiteral(node.property) && 'value' in node.property && node.property.value === 'execCommand'))) {
                    context.report({ node: node, messageId: 'execCommand' });
                }
            }
        };
    }
};
export default rule;
