// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-br-on-return.js
/**
 * @file Detect the usage of document.queryCommandSupported with 'insertBrOnReturn' argument
 */
import { isIdentifier, isCall, isLiteral, createDocumentBasedRuleVisitors } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            insertBrOnReturn: "insertBrOnReturn is not allowed since it is a Mozilla specific method, other browsers don't support that."
        },
        schema: []
    },
    create(context) {
        const createVisitors = createDocumentBasedRuleVisitors({
            isInteresting: (node, isDocumentObject) => {
                return (isCall(node.parent) &&
                    isDocumentObject(node.object) &&
                    ((isIdentifier(node.property) &&
                        node.property.name === 'queryCommandSupported') ||
                        (isLiteral(node.property) && node.property.value === 'queryCommandSupported')));
            },
            isValid: (node) => {
                return (node.parent.arguments.length === 0 ||
                    node.parent.arguments[0].value !== 'insertBrOnReturn');
            },
            messageId: 'insertBrOnReturn'
        });
        return createVisitors(context);
    }
};
export default rule;
