// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-global-selection.js
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
import { isIdentifier, contains, createIsWindowObject, createRememberWindow } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            globalSelection: 'Global selection modification, only modify local selections'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const FORBIDDEN_METHODS = ['getSelection'];
        // Create helper functions using the utility functions
        const isWindowObject = createIsWindowObject(WINDOW_OBJECTS);
        const rememberWindow = createRememberWindow(WINDOW_OBJECTS, isWindowObject);
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator'(node) {
                return rememberWindow(node.id, node.init);
            },
            'AssignmentExpression'(node) {
                return rememberWindow(node.left, node.right);
            },
            'MemberExpression'(node) {
                if (node &&
                    isWindowObject(node.object) &&
                    isIdentifier(node.property) &&
                    'name' in node.property &&
                    contains(FORBIDDEN_METHODS, node.property.name)) {
                    context.report({ node: node, messageId: 'globalSelection' });
                }
            }
        };
    }
};
export default rule;
