// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-dom-access.js
/**
 * @file Detect some warning for usages of (window.)document APIs
 */
import { isIdentifier, contains, createDocumentBasedRuleVisitors } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            domAccess: 'Direct DOM access, use jQuery selector instead'
        },
        schema: []
    },
    create(context) {
        const FORBIDDEN_DOCUMENT_METHODS = [
            'getElementById',
            'getElementsByClassName',
            'getElementsByName',
            'getElementsByTagName'
        ];
        const createVisitors = createDocumentBasedRuleVisitors({
            isInteresting: (node, isDocumentObject) => {
                return node && isDocumentObject(node.object);
            },
            isValid: (node) => {
                return !(isIdentifier(node.property) &&
                    contains(FORBIDDEN_DOCUMENT_METHODS, node.property.name));
            },
            messageId: 'domAccess'
        });
        return createVisitors(context);
    }
};
export default rule;
