// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-copy-to-clipboard.js
import { COPY_TO_CLIPBOARD } from '../language/diagnostics.js';
import { createFioriRule } from '../language/rule-factory.js';
import { createJsonFixer } from '../language/rule-fixer.js';
import { checkAppTablesConfiguration } from '../utils/helpers.js';
const rule = createFioriRule({
    ruleId: COPY_TO_CLIPBOARD,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Sets whether to enable or disable the copy to clipboard button in tables',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-copy-to-clipboard.md'
        },
        messages: {
            [COPY_TO_CLIPBOARD]: 'Copy To Clipboard in the {{sectionText}}table must be correctly configured. If not set, the "Copy" button is displayed'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            const parsedService = context.sourceCode.projectContext.getIndexedServiceForMainService(parsedApp);
            if (!parsedService) {
                continue;
            }
            for (const page of app.pages) {
                problems.push(...checkAppTablesConfiguration(page, parsedApp, checkConfiguration));
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: COPY_TO_CLIPBOARD,
            data: {
                sectionText: diagnostic.pageSectionName ? `${diagnostic.pageSectionName} ` : ''
            },
            fix: createJsonFixer({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
/**
 *
 * @param table
 * @returns
 */
function isV2Table(table) {
    return 'copy' in table.configuration;
}
/**
 *
 * @param page
 * @param table
 * @param parsedApp
 * @param problems
 * @param pageSectionName
 */
function checkConfiguration(page, table, parsedApp, problems, pageSectionName) {
    let config;
    let wrongValue = false;
    if (isV2Table(table)) {
        config = table.configuration.copy;
    }
    else {
        config = table.configuration.disableCopyToClipboard;
        wrongValue = true;
    }
    if (config?.valueInFile === wrongValue) {
        const copyIssue = {
            type: COPY_TO_CLIPBOARD,
            pageName: page.targetName,
            pageSectionName,
            manifest: {
                uri: parsedApp.manifest.manifestUri,
                object: parsedApp.manifestObject,
                propertyPath: config.configurationPath
            }
        };
        problems.push(copyIssue);
    }
}
export default rule;
