// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-dynamic-style-insertion.js
/**
 * @file Detect usage of document.styleSheets
 */
import { isIdentifier, isMember, isLiteral, createIsWindowObject, createRememberWindow, createIsDocument, createIsDocumentObject, createRememberDocument } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Rule Disablement
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Invoking global form of strict mode syntax for whole script
// ------------------------------------------------------------------------------
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            dynamicStyleInsertion: 'Dynamic style insertion, use library CSS or lessifier instead'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const DOCUMENT_OBJECTS = [];
        // Initialize factory functions
        const isWindowObject = createIsWindowObject(WINDOW_OBJECTS);
        const rememberWindow = createRememberWindow(WINDOW_OBJECTS, isWindowObject);
        const isDocument = createIsDocument(isWindowObject);
        const isDocumentObject = createIsDocumentObject(DOCUMENT_OBJECTS, isDocument);
        const rememberDocument = createRememberDocument(DOCUMENT_OBJECTS, isDocumentObject);
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a node represents an interesting dynamic style insertion.
         *
         * @param node The AST node to check
         * @returns True if the node represents dynamic style insertion, false otherwise
         */
        function isInteresting(node) {
            if (isMember(node) && isMember(node.object) && isDocumentObject(node.object.object)) {
                const prop = node.object.property;
                if (isIdentifier(prop) && prop.name === 'styleSheets') {
                    return true;
                }
                if (isLiteral(prop) && prop.value === 'styleSheets') {
                    return true;
                }
            }
            return false;
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init) || rememberDocument(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right) || rememberDocument(node.left, node.right);
            },
            'MemberExpression': function (node) {
                if (isInteresting(node)) {
                    context.report({ node: node, messageId: 'dynamicStyleInsertion' });
                }
            }
        };
    }
};
export default rule;
