// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-forbidden-window-property.js
/**
 * @file Detect the definition of global properties in the window object
 */
import { isIdentifier, isMember, isLiteral } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * Check if a node represents the global window variable.
 *
 * @param node The AST node to check
 * @returns True if the node represents the global window variable
 */
function isWindow(node) {
    // true if node is the global variable 'window'
    return !!(isIdentifier(node) && node && 'name' in node && node.name === 'window');
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            forbiddenWindowProperty: 'Usage of a forbidden window property.'
        },
        schema: []
    },
    create(context) {
        const WINDOW_OBJECTS = [];
        const FORBIDDEN_PROPERTIES = new Set(['top', 'addEventListener']);
        // --------------------------------------------------------------------------
        // Basic Helpers
        // --------------------------------------------------------------------------
        /**
         * Check if a node represents a window object or reference to it.
         *
         * @param node The AST node to check
         * @returns True if the node represents a window object or reference to it
         */
        function isWindowObject(node) {
            // true if node is the global variable 'window' or a reference to it
            return !!(isWindow(node) ||
                (node && isIdentifier(node) && 'name' in node && WINDOW_OBJECTS.includes(node.name)));
        }
        // --------------------------------------------------------------------------
        // Helpers
        // --------------------------------------------------------------------------
        /**
         * Remember window object references for later analysis.
         *
         * @param left The left side of the assignment
         * @param right The right side of the assignment
         * @returns True if window object was remembered, false otherwise
         */
        function rememberWindow(left, right) {
            if (isWindowObject(right) && isIdentifier(left) && 'name' in left) {
                WINDOW_OBJECTS.push(left.name);
                return true;
            }
            return false;
        }
        /**
         * Check if a node represents an interesting window property access.
         *
         * @param node The AST node to check
         * @returns True if the node represents an interesting window property access
         */
        function isInteresting(node) {
            return isMember(node) && isWindowObject(node.object);
        }
        /**
         * Check if a window property access is valid (not forbidden).
         *
         * @param node The AST node to validate
         * @returns True if the window property access is valid
         */
        function isValid(node) {
            let method = '';
            if (isIdentifier(node.property) && 'name' in node.property) {
                method = node.property.name;
            }
            if (isLiteral(node.property) && 'value' in node.property) {
                method = node.property.value;
            }
            return !FORBIDDEN_PROPERTIES.has(method);
        }
        // --------------------------------------------------------------------------
        // Public
        // --------------------------------------------------------------------------
        return {
            'VariableDeclarator': function (node) {
                return rememberWindow(node.id, node.init);
            },
            'AssignmentExpression': function (node) {
                return rememberWindow(node.left, node.right);
            },
            'MemberExpression': function (node) {
                if (isInteresting(node) && !isValid(node)) {
                    context.report({ node: node, messageId: 'forbiddenWindowProperty' });
                }
            }
        };
    }
};
export default rule;
