// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-enable-export.js
import { ENABLE_EXPORT } from '../language/diagnostics.js';
import { createFioriRule } from '../language/rule-factory.js';
import { createJsonFixer } from '../language/rule-fixer.js';
import { checkAppTablesConfiguration } from '../utils/helpers.js';
const rule = createFioriRule({
    ruleId: ENABLE_EXPORT,
    meta: {
        type: 'suggestion',
        docs: {
            recommended: true,
            description: 'Sets whether to enable or disable the "Export" button in tables',
            url: 'https://github.com/SAP/open-ux-tools/blob/main/packages/eslint-plugin-fiori-tools/docs/rules/sap-enable-export.md'
        },
        messages: {
            [ENABLE_EXPORT]: 'Export functionality in the {{sectionText}}table must be enabled'
        },
        fixable: 'code'
    },
    check(context) {
        const problems = [];
        for (const [appKey, app] of Object.entries(context.sourceCode.projectContext.linkedModel.apps)) {
            const parsedApp = context.sourceCode.projectContext.index.apps[appKey];
            const parsedService = context.sourceCode.projectContext.getIndexedServiceForMainService(parsedApp);
            if (!parsedService) {
                continue;
            }
            if (app.type === 'fe-v4') {
                for (const page of app.pages) {
                    problems.push(...checkAppTablesConfiguration(page, parsedApp, checkConfiguration));
                }
            }
        }
        return problems;
    },
    createJsonVisitorHandler: (context, diagnostic, deepestPathResult) => function report(node) {
        context.report({
            node,
            messageId: ENABLE_EXPORT,
            data: { sectionText: diagnostic.pageSectionName ? `${diagnostic.pageSectionName} ` : '' },
            fix: createJsonFixer({
                context,
                deepestPathResult,
                node,
                operation: 'delete'
            })
        });
    }
});
/**
 *
 * @param page
 * @param table
 * @param parsedApp
 * @param problems
 * @param pageSectionName
 */
function checkConfiguration(page, table, parsedApp, problems, pageSectionName) {
    if (table.configuration.enableExport.valueInFile === false) {
        problems.push({
            type: ENABLE_EXPORT,
            pageName: page.targetName,
            pageSectionName,
            manifest: {
                uri: parsedApp.manifest.manifestUri,
                object: parsedApp.manifestObject,
                propertyPath: table.configuration.enableExport.configurationPath
            }
        });
    }
}
export default rule;
