// @sap-ux/eslint-plugin-fiori-tools/lib/rules/sap-no-commons-usage.js
/**
 * @file Detects the usage of sap.ui.commons objects.
 */
import { isMember, isLiteral, isArray, startsWith, getMemberAsString } from '../utils/helpers.js';
// ------------------------------------------------------------------------------
// Helper Functions
// ------------------------------------------------------------------------------
/**
 * Check if a node represents an interesting function call to analyze.
 *
 * @param node The AST node to check
 * @returns True if the node represents an interesting function call
 */
function isInteresting(node) {
    const callee = node.callee;
    if (isMember(callee)) {
        if (getMemberAsString(callee) === 'sap.ui.define') {
            return true;
        }
    }
    return false;
}
/**
 * Check if a function call has valid (non-commons) imports.
 *
 * @param node The function call node to validate
 * @returns True if the function call has valid imports, false if it contains commons usage
 */
function isValid(node) {
    const nodeWithArgs = node;
    if (nodeWithArgs.arguments && isArray(nodeWithArgs.arguments[0])) {
        const importList = nodeWithArgs.arguments[0].elements;
        for (const key in importList) {
            if (importList.hasOwnProperty(key)) {
                const lib = importList[key];
                if (isLiteral(lib) && startsWith(lib.value, 'sap/ui/commons')) {
                    return false;
                }
            }
        }
    }
    return true;
}
// ------------------------------------------------------------------------------
// Rule Definition
// ------------------------------------------------------------------------------
const rule = {
    meta: {
        type: 'problem',
        docs: {
            description: 'fiori tools (fiori custom) ESLint rule',
            recommended: false
        },
        messages: {
            commonsUsage: 'Usage of sap.ui.commons controls is forbidden, please use controls from sap.m'
        },
        schema: []
    },
    create(context) {
        return {
            'NewExpression'(node) {
                if (isMember(node.callee) &&
                    startsWith(getMemberAsString(node.callee), 'sap.ui.commons')) {
                    context.report({ node: node, messageId: 'commonsUsage' });
                }
            },
            'CallExpression'(node) {
                if (isInteresting(node) && !isValid(node)) {
                    context.report({ node: node, messageId: 'commonsUsage' });
                }
            }
        };
    }
};
export default rule;
