// @sap-ux/fiori-annotation-api/dist/types/change.js
export var ExpressionType;
(function (ExpressionType) {
    ExpressionType["String"] = "String";
    ExpressionType["Bool"] = "Bool";
    ExpressionType["Decimal"] = "Decimal";
    ExpressionType["Date"] = "Date";
    ExpressionType["DateTimeOffset"] = "DateTimeOffset";
    ExpressionType["Float"] = "Float";
    ExpressionType["Guid"] = "Guid";
    ExpressionType["Int"] = "Int";
    ExpressionType["Path"] = "Path";
    ExpressionType["PropertyPath"] = "PropertyPath";
    ExpressionType["AnnotationPath"] = "AnnotationPath";
    ExpressionType["NavigationPropertyPath"] = "NavigationPropertyPath";
    ExpressionType["EnumMember"] = "EnumMember";
    ExpressionType["Collection"] = "Collection";
    ExpressionType["Record"] = "Record";
    ExpressionType["Unknown"] = "Unknown";
    ExpressionType["Null"] = "Null";
})(ExpressionType || (ExpressionType = {}));
export const MOD_TYPE_PRIMITIVE = 'primitive';
export const MOD_TYPE_EXPRESSION = 'expression';
export const MOD_TYPE_PROP_VALUE = 'property-value';
export const MOD_TYPE_RECORD = 'record';
export const MOD_TYPE_COLLECTION = 'collection';
export const MOD_TYPE_ANNOTATION = 'annotation';
export const MOD_TYPE_EMBEDDED_ANNOTATION = 'embedded-annotation';
export var ChangeType;
(function (ChangeType) {
    ChangeType["InsertAnnotation"] = "insert-annotation";
    ChangeType["InsertEmbeddedAnnotation"] = "insert-embedded-annotation";
    /**
     * Path should point to collection entry, property value or embedded annotation
     */
    ChangeType["Insert"] = "insert";
    /**
     * Path should point to collection entry, property value or embedded annotation
     */
    ChangeType["Delete"] = "delete";
    /**
     * Existing path target will be replaced; formatting/comments of original target gets lost -> only use for primitive values
     */
    ChangeType["Update"] = "update";
    /**
     * Move elements in collection
     */
    ChangeType["Move"] = "Move"; // path should point to collection entry that is moved
})(ChangeType || (ChangeType = {}));
